// @sap/ux-specification/dist/specification/src/sync/v4/export/pages/ListReport.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListReport = void 0;
const controls_1 = require("../controls");
const decorators_1 = require("../../../common/decoration/decorators");
const decorator_paths_1 = require("../../../common/decoration/decorator-paths");
const application_1 = require("../../application");
/**
 * Sync rule deleting all views if the chart view gets deleted.
 *
 * @param manifestSection - manifest section of the given view
 */
function deleteAllALPViews(manifestSection) {
    if (manifestSection['paths']?.[0]?.primary || manifestSection['paths']?.[0]?.secondary) {
        Object.keys(manifestSection).forEach((key) => delete manifestSection[key]);
    }
}
/**
 * Sync rule for preventing the export of a property.
 */
function doNotExport() {
    return;
}
/**
 * Sync rule for preventing the import of a property.
 */
function doNotImport() {
    return;
}
/**
 * Sync class for import and export of V4 list report.
 *
 * Extends `Decorator` so that `@hide` / `@message` decorators on properties (e.g. `inlineEdit`)
 * are evaluated when `new ListReport(settings).init()` is called during schema generation.
 *
 * Uses `app` context for app-level capabilities (FCL, collaborative draft)
 * and `page` context for page-level state (multi entity set, ALP, inlineEdit configured).
 *
 * @see app.usesFCL - set in App.config
 * @see app.usesCollaborativeDraft - set in App.config
 * @see page.hasMultipleEntitySets - set via page.addConfig() in ListReportPage constructor
 * @see page.isALP - set via page.addConfig() in ListReportPage constructor
 * @see page.inlineEditEnabled - set via page.addConfig() in ListReportPage constructor
 */
let ListReport = class ListReport {
};
exports.ListReport = ListReport;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName) => (0, application_1.getSettingsPathForPage)(pageName) + '/views',
            import: doNotImport,
            export: doNotExport,
            delete: deleteAllALPViews
        }
    })
], ListReport.prototype, "chart", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName) => (0, application_1.getSettingsPathForPage)(pageName)
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], ListReport.prototype, "variantManagement", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName) => (0, application_1.getSettingsPathForPage)(pageName) + '/views/paths/[]/0'
        }
    }),
    (0, decorators_1.descriptionSrcURL)('https://ui5.sap.com/sdk/#/topic/2a9df06673d34f72b238549d49da8bfb')
], ListReport.prototype, "defaultPath", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName) => (0, application_1.getSettingsPathForPage)(pageName)
        }
    }),
    (0, decorators_1.hide)(true)
], ListReport.prototype, "defaultTemplateAnnotationPath", void 0);
__decorate([
    (0, decorators_1.childId)('header')
], ListReport.prototype, "header", void 0);
__decorate([
    (0, decorators_1.childId)('fe::FilterBar')
], ListReport.prototype, "filterBar", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName) => (0, application_1.getSettingsPathForPage)(pageName),
            import: controls_1.importInlineEdit,
            export: controls_1.exportInlineEdit
        }
    }),
    (0, decorators_1.descriptionSrcURL)('https://ui5.sap.com/sdk/#/topic/bb56175f37aa4530a59214025a005466'),
    (0, decorators_1.validity)({
        since: '1.136.0'
    }),
    (0, decorators_1.hide)((0, decorator_paths_1.and)((0, decorator_paths_1.or)(decorator_paths_1.page.isALP, decorator_paths_1.app.usesFCL, decorator_paths_1.page.hasMultipleEntitySets, decorator_paths_1.app.usesCollaborativeDraft), (0, decorator_paths_1.not)(decorator_paths_1.page.inlineEditEnabled))),
    (0, decorators_1.message)((0, decorator_paths_1.and)(decorator_paths_1.app.usesCollaborativeDraft, decorator_paths_1.page.inlineEditEnabled), (0, decorator_paths_1.msg)('INLINE_EDIT_NOT_SUPPORTED_FOR_COLLABORATIVE_DRAFT', {}, { deletable: true })),
    (0, decorators_1.message)((0, decorator_paths_1.and)(decorator_paths_1.page.isALP, decorator_paths_1.page.inlineEditEnabled), (0, decorator_paths_1.msg)('INLINE_EDIT_NOT_SUPPORTED_FOR_ANALYTICAL_LIST_PAGE', {}, { deletable: true })),
    (0, decorators_1.message)((0, decorator_paths_1.and)(decorator_paths_1.app.usesFCL, decorator_paths_1.page.inlineEditEnabled), (0, decorator_paths_1.msg)('INLINE_EDIT_NOT_SUPPORTED_FOR_FCL', {}, { deletable: true })),
    (0, decorators_1.message)((0, decorator_paths_1.and)(decorator_paths_1.page.hasMultipleEntitySets, decorator_paths_1.page.inlineEditEnabled), (0, decorator_paths_1.msg)('INLINE_EDIT_NOT_SUPPORTED_FOR_MULTI_ENTITY_SET', {}, { deletable: true }))
], ListReport.prototype, "inlineEdit", void 0);
exports.ListReport = ListReport = __decorate([
    (0, decorators_1.pageLayoutInformation)({
        id: 'sap.suite.ui.generic.template.ListReport.view.ListReport',
        pageId: 'sap.suite.ui.generic.template.ListReport'
    })
], ListReport);
