// @sap/ux-specification/dist/specification/src/ftfs/parser/model/types/annotations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UI_NODE_TYPE_VISUAL_FILTER = exports.UI_NODE_TYPE_VISUAL_FILTERS = exports.UI_NODE_TYPE_FILTER_FIELD = exports.UI_NODE_TYPE_FILTER_FIELDS = exports.UI_NODE_TYPE_COLUMN = exports.UI_NODE_TYPE_LIST = exports.UI_NODE_TYPE_VIEWS = exports.UI_NODE_TYPE_NAVIGATION = exports.UI_NODE_TYPE_ANALYTICAL_CHART = exports.UI_NODE_TYPE_DATA_POINT = exports.UI_NODE_TYPE_CONTACT = exports.UI_NODE_TYPE_CHART = exports.UI_NODE_TYPE_GENERIC_ACTION = exports.UI_NODE_TYPE_GENERIC_ACTIONS = exports.UI_NODE_TYPE_ACTION_GROUP = exports.UI_NODE_TYPE_ACTION = exports.UI_NODE_TYPE_CONNECTED_FIELDS = exports.UI_NODE_TYPE_FORM_FIELD = exports.UI_NODE_TYPE_SECTION = exports.UI_NODE_TYPE_HEADER_SECTIONS = exports.UI_NODE_TYPE_FORM = exports.UI_NODE_TYPE_PAGE_ACTIONS = exports.UI_NODE_TYPE_HEADER_INFO = void 0;
// UI Node Types (OP)
exports.UI_NODE_TYPE_HEADER_INFO = 'headerInfo';
exports.UI_NODE_TYPE_PAGE_ACTIONS = 'pageActions';
exports.UI_NODE_TYPE_FORM = 'form';
exports.UI_NODE_TYPE_HEADER_SECTIONS = 'headerSections';
exports.UI_NODE_TYPE_SECTION = 'section';
exports.UI_NODE_TYPE_FORM_FIELD = 'formField';
exports.UI_NODE_TYPE_CONNECTED_FIELDS = 'connectedFields';
exports.UI_NODE_TYPE_ACTION = 'action';
exports.UI_NODE_TYPE_ACTION_GROUP = 'actionGroup';
exports.UI_NODE_TYPE_GENERIC_ACTIONS = 'genericActions';
exports.UI_NODE_TYPE_GENERIC_ACTION = 'genericAction';
exports.UI_NODE_TYPE_CHART = 'chart';
exports.UI_NODE_TYPE_CONTACT = 'contact';
exports.UI_NODE_TYPE_DATA_POINT = 'dataPoint';
exports.UI_NODE_TYPE_ANALYTICAL_CHART = 'analyticalChart';
exports.UI_NODE_TYPE_NAVIGATION = 'navigation';
// UI Node Types (LR)
exports.UI_NODE_TYPE_VIEWS = 'views'; // for multi view LR: parent node containing views listed in manifest json (views/paths) - could be of type 'list' or 'analyticalChart'
exports.UI_NODE_TYPE_LIST = 'list';
exports.UI_NODE_TYPE_COLUMN = 'column';
exports.UI_NODE_TYPE_FILTER_FIELDS = 'filterFields';
exports.UI_NODE_TYPE_FILTER_FIELD = 'filterField';
exports.UI_NODE_TYPE_VISUAL_FILTERS = 'visualFilters';
exports.UI_NODE_TYPE_VISUAL_FILTER = 'visualFilter';
