// @sap/ux-specification/dist/specification/src/sync/v2/genericSchemaHandling/controls/MultiEdit.js
"use strict";
/**
 * @file This file is a helper class for its sibling class Table.ts. It provides the logic needed to support the multiEdit property (in LR and ALP).
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.syncRuleMultiEdit = exports.MultiEdit = void 0;
const i18next_1 = __importDefault(require("i18next"));
const v2 = __importStar(require("@sap/ux-specification-types/src/v2/genericSchemaHandling/controls/Table"));
const decoration_1 = require("../../../common/decoration");
const Table_1 = require("./Table");
const utils_1 = require("../../utils");
const manifestPropertyUtils_1 = require("../../generate/manifestPropertyUtils");
const common_1 = require("../../../common");
/**
 * This function adapts getSyncRuleForManifestProperty for properties belonging to the multiEdit object inside the tableSettings.
 *
 * @param property - the property name inside multiEdit
 * @param propertyIndex - index of the property inside multiEdit (used to achieve an ordering which is more logical than alphabetical)
 * @param expectedType -type of the property
 * @param onlyEnabled - whether the property is only allowed when multiEdit is enabled
 * @param addEnum - optional function for adding enum values to the property definition
 * @returns the SyncRule for the multiEdit property
 */
function getSyncRuleForMultiEditProperty(property, propertyIndex, expectedType, onlyEnabled = false, addEnum) {
    const analyze = function (element, schemaHandlingParams) {
        const multiEditInfo = schemaHandlingParams.specificParams;
        const rule = {
            manifestPath: multiEditInfo.manifestPath
        };
        if (onlyEnabled && !multiEditInfo.enabled) {
            rule.disallowedReason = i18next_1.default.t('PROPERTYONLYMULTIEDIT', { property });
        }
        return rule;
    }; // analyze
    const adapt = function (element, schemaHandlingParams, generateParameters) {
        element['propertyIndex'] = propertyIndex;
        if (addEnum) {
            const multiEditInfo = schemaHandlingParams.specificParams;
            if (multiEditInfo.entityTypeDefinition) {
                addEnum(element, multiEditInfo.entityTypeDefinition);
            }
        }
    };
    return (0, manifestPropertyUtils_1.getSyncRuleForManifestProperty)(property, analyze, adapt, undefined, expectedType);
}
// MultiEdit
// This is the type which is used for describing the multiEdit manifest property (only in case this property is allowed for the current table)
let MultiEdit = class MultiEdit {
};
exports.MultiEdit = MultiEdit;
__decorate([
    (0, decoration_1.syncRule)(getSyncRuleForMultiEditProperty('enabled', 0, 'boolean'))
], MultiEdit.prototype, "enabled", void 0);
__decorate([
    (0, decoration_1.syncRule)(getSyncRuleForMultiEditProperty('annotationPath', 1, 'string', true, function (element, entityTypeDefinition) {
        const fieldGroups = (0, common_1.getListOfValidAnnotations)(entityTypeDefinition, ["com.sap.vocabularies.UI.v1.FieldGroup" /* UIAnnotationTerms.FieldGroup */]);
        element.enum = fieldGroups;
    }))
], MultiEdit.prototype, "annotationPath", void 0);
__decorate([
    (0, decoration_1.syncRule)(getSyncRuleForMultiEditProperty('ignoredFields', 2, 'string', true, function (bluePrint, entityTypeDefinition) {
        (0, Table_1.addEntityTypeFieldsAsEnum)(bluePrint, entityTypeDefinition);
    }))
], MultiEdit.prototype, "ignoredFields", void 0);
exports.MultiEdit = MultiEdit = __decorate([
    (0, decoration_1.syncRule)({
        processingRuleAdapter(processingRule, schemaHandlingParams, generateParameters) {
            const tableParams = schemaHandlingParams.specificParams;
            const settings = tableParams.multiTableInfo?.variantInfo
                ? tableParams.multiTableInfo.variantInfo.variantDefinition
                : tableParams.settings;
            const tableSettings = settings?.['tableSettings'];
            const hasTypeError = (0, manifestPropertyUtils_1.hasPropertyTypeError)(tableSettings, 'multiEdit', generateParameters.logger, 'object');
            if (hasTypeError) {
                delete processingRule.element;
                return;
            }
            const multiEdit = tableSettings?.multiEdit;
            const enabled = !!multiEdit?.enabled;
            const entityTypeDefinition = tableParams.multiTableInfo?.variantInfo
                ? tableParams.multiTableInfo.variantInfo.entityTypeDefinition
                : tableParams.entityType;
            const manifestPath = (0, Table_1.getManifestPathForTableSettings)(schemaHandlingParams, tableParams);
            manifestPath.push('multiEdit');
            const multiEditInfo = {
                enabled,
                multiEdit,
                entityTypeDefinition,
                manifestPath
            };
            processingRule.referenceAdaptation = {
                specificParams: multiEditInfo
            };
        }
    })
], MultiEdit);
exports.syncRuleMultiEdit = {
    processingRuleAdapter(processingRule, schemaHandlingParams, generateParameters) {
        const tableParams = schemaHandlingParams.specificParams;
        const tableType = (0, Table_1.getTableType)(tableParams);
        let disallowedReason;
        if (tableParams.settings?.isWorklist) {
            disallowedReason = i18next_1.default.t('MULTIEDITNOWORKLIST');
        }
        else if (tableParams.multiTableInfo?.variantInfo?.isChart) {
            disallowedReason = i18next_1.default.t('MULTIEDITNOCHART');
        }
        else if (tableType === v2.TableTypeV2Enum.AnalyticalTable || tableType === v2.TableTypeV2Enum.TreeTable) {
            disallowedReason = i18next_1.default.t('MULTIEDITTABLETYPE');
        }
        if (disallowedReason) {
            const manifestPath = (0, Table_1.getManifestPathForTableSettings)(schemaHandlingParams, tableParams);
            const { value } = (0, utils_1.getManifestPropertyByPath)(generateParameters.manifest, manifestPath, 'multiEdit');
            (0, manifestPropertyUtils_1.handleDisallowedProperty)(disallowedReason, processingRule, value, manifestPath);
            return;
        }
        processingRule.referenceAdaptation = {
            suffix: tableParams.multiTableInfo?.variantInfo
                ? tableParams.multiTableInfo.variantInfo.iconTabFilterKey
                : undefined,
            syncRuleProvider: MultiEdit
        };
    }
};
