// @sap/ux-specification/dist/specification/src/sync/v2/genericSchemaHandling/controls/TableColumn.js
"use strict";
/**
 * @file This file is a helper module for its sibling module Table.ts. It provides the logic needed to support the columns of a table.
 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.syncRuleColumns = exports.GeneralTableColumns = exports.CustomColumnDefinition = exports.GeneralTableColumn = void 0;
const decoration_1 = require("../../../common/decoration");
const common_1 = require("../../../common");
const ux_specification_types_1 = require("@sap/ux-specification-types");
const stableIdUtils_1 = require("../../generate/stableIdUtils");
const Action_1 = require("./Action");
const Table_1 = require("./Table");
const utils_1 = require("../../generate/utils");
/**
 * The FlexAdapter used for the flex based properties of columns. Provides a control type that will be written into the corresponding flex changes.
 * Currently, this is always sap.m.Column. Note that this is actually only true for responsive tables. For other tables it would be sap.ui.table.Column.
 * Hence, changing the table type should result in an update of all flex changes for columns.
 * Currently, we do not have a logic that supports that.
 * Luckily, UI5 does not evaluate this property of a flex change, since the control id anyway identifies the control.
 */
const flexRuleForTableColumn = {
    controlType: () => ux_specification_types_1.ControlType.TableColumn
};
/**
 * The ProcessingRuleAdapter for the flex based properties of columns.
 *
 * @param processingRule - the ProcessingRule to be adapted
 * @param schemaHandlingParams - information provided by the hierarchy level above
 */
const processingRuleAdapterForTableColumnFlexProperty = function (processingRule, schemaHandlingParams) {
    const columnDetail = schemaHandlingParams.specificParams;
    if (columnDetail.controlId) {
        processingRule.controlId = columnDetail.controlId;
    }
    else {
        // As the column has no stable id flex based properties should be suppressed.
        delete processingRule.element;
    }
};
/**
 * Provides a sync rule for a flex based property of a column.
 *
 * @param supportsExpressionBindings - information whether the option for expression bindings should be added
 * @returns the syncRule to be used by the property
 */
function getSyncRuleForTableColumnProperty(supportsExpressionBindings) {
    const syncRule = {
        flex: flexRuleForTableColumn,
        processingRuleAdapter: processingRuleAdapterForTableColumnFlexProperty
    };
    if (supportsExpressionBindings) {
        syncRule.generate = decoration_1.addPatternForBindingChangeOfEnumLR; // add option for bindings
    }
    return syncRule;
}
// GeneralTableColumn
let GeneralTableColumn = class GeneralTableColumn {
};
exports.GeneralTableColumn = GeneralTableColumn;
__decorate([
    (0, decoration_1.syncRule)(getSyncRuleForTableColumnProperty(false))
], GeneralTableColumn.prototype, "width", void 0);
__decorate([
    (0, decoration_1.syncRule)(getSyncRuleForTableColumnProperty(true))
], GeneralTableColumn.prototype, "hAlign", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        processingRuleAdapter(processingRule, schemaHandlingParams) {
            const columnDetail = schemaHandlingParams.specificParams;
            if (columnDetail.actionInfo) {
                processingRule.referenceAdaptation = {
                    specificParams: columnDetail.actionInfo,
                    suffix: (0, stableIdUtils_1.getStableIdForDatafieldActionButton)(columnDetail.actionInfo),
                    syncRuleProvider: Action_1.Action
                };
            }
            else {
                delete processingRule.element; // no action info available -> suppress action node
            }
        }
    })
], GeneralTableColumn.prototype, "action", void 0);
exports.GeneralTableColumn = GeneralTableColumn = __decorate([
    (0, decoration_1.syncRule)({
        processingRuleAdapter(processingRule, schemaHandlingParams) {
            const element = processingRule.element;
            const columnInfo = schemaHandlingParams.specificParams;
            // Add some information to processingRule.element:
            const description = columnInfo.column.description;
            element.description = description;
            element[ux_specification_types_1.SchemaTag.dataType] = (0, common_1.determineDataType)(columnInfo.column.dataField);
            if (columnInfo.column.message) {
                element[ux_specification_types_1.SchemaTag.messages] = [columnInfo.column.message];
            }
            (0, common_1.addKeyToElement)(columnInfo.column.dataField, element, false);
            // Translate the specificParams from type ColumnInfo to type ColumnDetail which is easier to consume by the properties of class GeneralTableColumn:
            const controlId = (0, stableIdUtils_1.getControlIdForColumn)(columnInfo.smartTableId, columnInfo.column.dataField, columnInfo.column.isAction);
            let actionInfo;
            if (columnInfo.column.isAction) {
                actionInfo = {
                    dataField: columnInfo.column,
                    action: columnInfo.column.dataField,
                    iconTabFilterKey: columnInfo.iconTabFilterKey,
                    isViewNode: false
                };
            }
            const columnDetail = {
                controlId,
                actionInfo
            };
            processingRule.referenceAdaptation = {
                specificParams: columnDetail
            };
        }
    })
], GeneralTableColumn);
// Instances of this class represent custom columns.
// Note that the structure of this class is given by the specific implementation for custom columns in tools-suite.
class CustomColumnDefinition {
}
exports.CustomColumnDefinition = CustomColumnDefinition;
__decorate([
    (0, decoration_1.syncRule)({
    // todo: Implement when custom columns are supported.
    })
], CustomColumnDefinition.prototype, "leadingProperty", void 0);
__decorate([
    (0, decoration_1.syncRule)({
    // todo: Implement when custom columns are supported.
    })
], CustomColumnDefinition.prototype, "fragmentName", void 0);
__decorate([
    (0, decoration_1.syncRule)({
    // todo: Implement when custom columns are supported.
    })
], CustomColumnDefinition.prototype, "cellsFragmentName", void 0);
// CustomColumnDefinitions
let CustomColumnDefinitions = class CustomColumnDefinitions {
};
__decorate([
    (0, decoration_1.syncRule)({
    // todo: Implement when custom columns are supported.
    })
], CustomColumnDefinitions.prototype, "items", void 0);
CustomColumnDefinitions = __decorate([
    (0, decoration_1.syncRule)({
    // todo: Implement when custom columns are supported.
    })
], CustomColumnDefinitions);
/**
 * This function provides the additional 'custom' entry to be added to the list of columns representing the custom columns.
 *
 * @returns Optional manifest path and schema definition for custom columns; both are undefined when no custom columns should be generated.
 */
function getCustomColumnsDefinition() {
    return {
        manifestPath: undefined,
        customColumnsDefinition: undefined
    }; // Todo: Implement logic for custom columns
}
const ACCESSOR_CUSTOM_COLUMNS = 'custom';
// GeneralTableColumns
let GeneralTableColumns = class GeneralTableColumns {
};
exports.GeneralTableColumns = GeneralTableColumns;
exports.GeneralTableColumns = GeneralTableColumns = __decorate([
    (0, decoration_1.syncRule)({
        processingRuleAdapter(processingRule, schemaHandlingParams) {
            const columnsInfo = schemaHandlingParams.specificParams;
            const columnDefinitions = columnsInfo.columns;
            const properties = {};
            const columnInfos = {};
            columnDefinitions.forEach(function (columnDefinition, i) {
                const propertyDefinition = {
                    $ref: common_1.DEFINITION_LINK_PREFIX + 'GeneralTableColumn'
                };
                propertyDefinition[ux_specification_types_1.SchemaTag.propertyIndex] = i;
                // Note: The accessor defined below must be a unique identifier of the column within the list of columns.
                // On the first glance columnDefinition.dataField.stableIdPart looks like a good candidate for that.
                // However, there are edge cases in which there are two columns defined which have the same stableIdPart.
                // This indicates either a situation in which the page will fail to load (because of a duplicate id) or a situation
                // in which some column does not have a stable id at all. Fiori Tools should be able to handle the later case, but no flex changes should
                // be allowed for a column without stable id (which is ensured by syncRuleForTableColumnProperty).
                const accessor = `TableColumn${i}`;
                properties[accessor] = propertyDefinition;
                const columnInfo = {
                    // facet: columnsInfo.facet
                    column: columnDefinition,
                    smartTableId: columnsInfo.smartTableId,
                    iconTabFilterKey: columnsInfo.iconTabFilterKey
                };
                columnInfos[accessor] = columnInfo;
            });
            const { customColumnsDefinition } = getCustomColumnsDefinition();
            if (customColumnsDefinition) {
                properties[ACCESSOR_CUSTOM_COLUMNS] = customColumnsDefinition;
            }
            processingRule.element.properties = properties;
            processingRule.element.additionalProperties = false;
            processingRule.referenceAdaptation = {
                processingRulesAdapter(property, processingRule) {
                    const isCustomColumn = property === ACCESSOR_CUSTOM_COLUMNS;
                    const columnInfo = columnInfos[property];
                    processingRule.referenceAdaptation = {
                        suffix: `${isCustomColumn ? '' : property}${(0, stableIdUtils_1.getSuffixFromIconTabFilterKey)(columnsInfo.iconTabFilterKey)}`,
                        specificParams: columnInfo,
                        syncRuleProvider: isCustomColumn ? CustomColumnDefinitions : GeneralTableColumn
                    };
                }
            };
        }
    })
], GeneralTableColumns);
/**
 * The SyncRule for the columns of a table. It collects the information needed to generate the nodes for the columns and to adapt the nodes for the columns in the manifest and to provide the information needed to generate flex changes for the columns.
 */
exports.syncRuleColumns = {
    processingRuleAdapter(processingRule, schemaHandlingParams) {
        const tableParams = schemaHandlingParams.specificParams;
        if (tableParams.multiTableInfo &&
            (!tableParams.multiTableInfo.variantInfo || tableParams.multiTableInfo.variantInfo.isChart)) {
            delete processingRule.element;
            return; // no columns defined in the global table settings of a multi table app and for chart views
        }
        const lineItemInfo = tableParams.multiTableInfo
            ? tableParams.multiTableInfo.variantInfo.lineItemInfo
            : tableParams.lineItemInfo;
        if (!lineItemInfo) {
            delete processingRule.element; // no line item available -> suppress columns. Note that this can only happen in multi table case. If global lineItem is not available whole table node will be skipped.
            return;
        }
        const tableType = (0, Table_1.getTableType)(tableParams);
        const smartTableId = tableParams.multiTableInfo
            ? tableParams.multiTableInfo.variantInfo.smartControlId
            : tableParams.tableId;
        const columns = lineItemInfo.dataFieldInfos.filter(function (dataField) {
            if (dataField.target !== utils_1.DataFieldTarget.Column) {
                return false;
            }
            const dataFieldDefinition = dataField.dataField;
            if (dataFieldDefinition.$Type === "com.sap.vocabularies.UI.v1.DataField" /* UIAnnotationTypes.DataField */ ||
                dataFieldDefinition.$Type === "com.sap.vocabularies.UI.v1.DataFieldWithUrl" /* UIAnnotationTypes.DataFieldWithUrl */) {
                // These DataFields are excluded by the SmartTable if they do not have a Path.
                // Note: There seem to be situations when FE decides to create a column for a DataField without Path anyway (probably if they should host the DraftIndicator).
                // Currently, we ignore this edge case within an edge case.
                return !!dataFieldDefinition.Value?.['path'];
            }
            return true;
        });
        const entityType = tableParams.multiTableInfo
            ? tableParams.multiTableInfo.variantInfo.entityTypeDefinition
            : tableParams.entityType;
        const columnsInfo = {
            entityType,
            tableType,
            columns,
            smartTableId,
            iconTabFilterKey: tableParams.multiTableInfo?.variantInfo.iconTabFilterKey
        };
        processingRule.referenceAdaptation = {
            suffix: smartTableId,
            specificParams: columnsInfo,
            syncRuleProvider: GeneralTableColumns
        };
        processingRule.element[ux_specification_types_1.SchemaTag.propertyIndex] = 1; // place columns node behind the toolbar node (which has propertyIndex 0).
    }
};
