// @sap/ux-specification/dist/specification/src/ftfs/parser/model/actions/ActionGroupAggregation.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActionGroupAggregation = void 0;
const types_1 = require("../types");
const ActionAggregation_1 = require("./ActionAggregation");
const i18next_1 = __importDefault(require("i18next"));
const ActionsContainerAggregation_1 = require("./ActionsContainerAggregation");
const ActionsAggregation_1 = require("./ActionsAggregation");
class ActionGroupAggregation extends ActionsContainerAggregation_1.ActionsContainerAggregation {
    /**
     * Creates an instance of `ActionGroupAggregation`.
     * Reads `actionType` from the schema to set `menuType` and, for `CustomMenu` groups
     * that are not pure aggregations, registers a custom creation form.
     *
     * @param data Optional aggregation data object used to initialize properties.
     * @param schema Optional JSON schema fragment associated with this aggregation.
     */
    constructor(data, schema) {
        super(data, schema);
        this.sortableList = true;
        this.childClass = ActionAggregation_1.ActionAggregation;
        this.allowedAnnotationCreationForms = [];
        this.actions = [{ type: types_1.AggregationActions.Delete }];
        this.sortableItem = types_1.SortingOptions.Enabled;
        this.sortableCollection = 'actions';
        this.i18nKey = 'ACTION_MENU';
        this.disallowEmpty = true;
        this.menuType = schema?.actionType ?? '';
        // Custom creation form - check schema if supported
        if (this.schema?.properties &&
            this.schema?.actionType === 'CustomMenu' &&
            this.schema?.metadata?.type !== 'Aggregation') {
            this.schemaCreationForms = [
                {
                    name: types_1.AggregationCreationForm.CustomAction,
                    kind: types_1.SCHEMA_CREATION_FORM,
                    title: 'PAGE_EDITOR_OUTLINE_ADD_CUSTOM_ACTIONS_TITLE',
                    disabled: false
                }
            ];
            this.custom = true;
            this.additionalText = i18next_1.default.t('PAGE_EDITOR_OUTLINE_NODE_DESC_CUSTOM_ACTION_MENU');
        }
    }
    /**
     * Overwritten method for data update of object page actions
     * Method receives current values for actions and detects custom actions.
     *
     * @param data Data which should be used for value population.
     * @param page Page config data.
     * @param pageType Page type.
     * @param path Aggregation path.
     * @param annotations Annotations data.
     */
    updatePropertiesValues(data, page, pageType, path, annotations) {
        super.updatePropertiesValues(data, page, pageType, path, annotations);
        this.formSchema = this.actionsContainer.additionalProperties?.aggregations['actions'];
        const actions = data?.actions ?? {};
        if (actions && typeof actions === 'object') {
            for (const id in actions) {
                const action = this.aggregations['actions'].aggregations[id];
                if (action?.schema && (action.schema.actionType === 'Custom' || !action.schema.annotationPath)) {
                    action.markAsCustomAction();
                }
            }
        }
    }
    /**
     * Returns the internal actions container aggregation.
     * For custom action groups this is the nested `ActionsAggregation`; for annotation-based groups it is `this`.
     *
     * @returns The `ActionsAggregation` that holds the action items for this group.
     */
    get actionsContainer() {
        return this.custom ? this.aggregations['actions'] : this;
    }
    /**
     * Resolves the child class to use when instantiating nested aggregation nodes.
     * Custom action groups expose a nested `ActionsAggregation` under the `actions` key;
     * annotation-based action groups only accept children that carry an `annotationPath`.
     *
     * @param schema - The JSON schema of the property being resolved.
     * @param path - The property path leading to the current node.
     * @returns `ActionsAggregation` when the last path segment is `'actions'`, the default
     *   `childClass` when the schema has an `annotationPath`, or `undefined` otherwise.
     */
    getChildClass(schema, path = []) {
        if (path[path.length - 1] === 'actions') {
            return ActionsAggregation_1.ActionsAggregation; // Allow nested 'actions' aggregation for custom action groups
        }
        return schema?.annotationPath ? this.childClass : undefined; // annotation bases action groups don't have nested 'actions' aggregation
    }
    /**
     * Method adds aggregation object.
     * Overwritten to mark internal actions container as nested.
     *
     * @param name Name of aggregation.
     * @param aggregation Aggregation to add.
     * @param path Array of path to aggregation.
     * @param order Order index.
     * @param overwrite Overwrite existing aggregation.
     * @returns Added aggregation.
     */
    addAggregation(name, aggregation, path, order, overwrite) {
        aggregation = super.addAggregation(name, aggregation, path, order, overwrite);
        if (aggregation instanceof ActionsAggregation_1.ActionsAggregation) {
            aggregation.isNestedContainer = true;
        }
        return aggregation;
    }
    /**
     * Method returns array of ordered aggregations.
     * Overridden to return nested actions container aggregations for custom action groups.
     *
     * @param viewNodesOnly Return only aggregations for view.
     * @param sortBy Sorting type. Currently aggregations can be sorted by 'ViewNode' to move visible nodes to top.
     * @returns Array of aggregations.
     */
    getAggregations(viewNodesOnly = false, sortBy) {
        return this.custom && viewNodesOnly
            ? this.actionsContainer.getAggregations(viewNodesOnly, sortBy)
            : super.getAggregations(viewNodesOnly, sortBy);
    }
    /**
     * Returns the keys of all menu item actions belonging to this action group.
     * For custom action groups the items are sourced from the nested `ActionsAggregation`;
     * for annotation-based groups they are taken directly from `this`.
     *
     * @returns An array of aggregation key strings representing the menu items.
     */
    getMenuItems() {
        return (this.custom ? this.aggregations['actions'] : this).getAggregationKeys();
    }
}
exports.ActionGroupAggregation = ActionGroupAggregation;
