// @sap/ux-specification/dist/specification/src/ftfs/parser/model/actions/ActionsAggregation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActionsAggregation = void 0;
const types_1 = require("../types");
const ActionGroupAggregation_1 = require("./ActionGroupAggregation");
const ActionsContainerAggregation_1 = require("./ActionsContainerAggregation");
class ActionsAggregation extends ActionsContainerAggregation_1.ActionsContainerAggregation {
    /**
     * Creates an instance of `ActionsAggregation`.
     * Registers a custom creation form when the schema exposes `additionalProperties`
     * and is not a pure aggregation type.
     *
     * @param data Optional aggregation data object used to initialize properties.
     * @param schema Optional JSON schema fragment associated with this aggregation.
     */
    constructor(data, schema) {
        super(data, schema);
        this.allowedAnnotationCreationForms = [types_1.AggregationCreationForm.NativeAction, types_1.AggregationCreationForm.NativeNavigation];
        // Custom creation form - check schema if supported
        if (schema?.properties && this.schema?.additionalProperties && this.schema?.metadata?.type !== 'Aggregation') {
            this.schemaCreationForms = [
                {
                    name: types_1.AggregationCreationForm.CustomAction,
                    kind: types_1.SCHEMA_CREATION_FORM,
                    title: 'PAGE_EDITOR_OUTLINE_ADD_CUSTOM_ACTIONS_TITLE',
                    disabled: false
                }
            ];
        }
    }
    /**
     * Resolves the child class to use when instantiating a nested action node.
     * Returns `ActionGroupAggregation` for annotation-based action groups (`DataFieldForActionGroup`)
     * and custom menu groups (`CustomMenu`); falls back to the default `childClass` otherwise.
     *
     * @param schema - The JSON schema of the property being resolved.
     * @returns The appropriate child class constructor, or `undefined` if none is defined.
     */
    getChildClass(schema) {
        if (schema?.dataType === 'DataFieldForActionGroup' || schema?.actionType === 'CustomMenu') {
            return ActionGroupAggregation_1.ActionGroupAggregation;
        }
        return this.childClass;
    }
}
exports.ActionsAggregation = ActionsAggregation;
