// @sap/ux-specification/dist/specification/src/sync/common/decoration/steps/Enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Enums = void 0;
const decorators_1 = require("../decorators");
const condition_evaluator_1 = require("../condition-evaluator");
const enum_utils_1 = require("./enum-utils");
const decorator_paths_1 = require("../decorator-paths");
const jsonpath_plus_1 = require("jsonpath-plus");
/**
 * Resolves allowedValues from static array or PathNode reference.
 * Uses the same evaluation context as the original implementation.
 *
 * @param allowedValues - Static array or PathNode to resolve
 * @param context - Runtime context for path resolution
 * @returns Array of allowed enum values
 */
function resolveAllowedValues(allowedValues, context) {
    if ((0, decorator_paths_1.isPathNode)(allowedValues)) {
        const path = (0, decorator_paths_1.getPath)(allowedValues);
        const evaluationContext = {
            app: context.app,
            page: context.page,
            custom: context.custom
        };
        const result = (0, jsonpath_plus_1.JSONPath)({ path: `$.${path}`, json: evaluationContext });
        return Array.isArray(result) && result.length > 0 ? result[0] : [];
    }
    return allowedValues;
}
/**
 * Applies enum filtering to a single property if conditions match.
 *
 * @param property - Schema property definition
 * @param enumsMetadata - Enum metadata from decorator
 * @param evaluator - Condition evaluator instance
 * @param context - Runtime context
 * @returns True if filtering was applied, false otherwise
 */
function applyEnumFiltering(property, enumsMetadata, evaluator, context) {
    const { passed } = evaluator.evaluate(enumsMetadata);
    if (!passed) {
        return false;
    }
    const currentEnumValues = (0, enum_utils_1.resolveEnumFromProperty)(property, context);
    if (!currentEnumValues) {
        return false;
    }
    const allowedValues = resolveAllowedValues(enumsMetadata.allowedValues, context);
    const filteredValues = allowedValues.filter((v) => currentEnumValues.includes(v));
    (0, enum_utils_1.applyFilteredEnumToProperty)(property, filteredValues, context);
    return true;
}
/**
 * Enums step - filters enum values based on conditions.
 * If multiple @enums decorators exist on the same property, the first matching condition wins.
 */
class Enums {
    constructor() {
        this.name = 'enums';
    }
    /**
     * Applies enum filtering to all properties in the definition.
     *
     * @param target - The class prototype containing decorator metadata
     * @param definition - The schema definition to modify
     * @param context - Runtime context for condition evaluation
     */
    init(target, definition, context) {
        if (!definition?.properties || !target) {
            return;
        }
        const evaluator = new condition_evaluator_1.ConditionEvaluator(context);
        for (const propertyName in definition.properties) {
            const property = definition.properties[propertyName];
            const enumsMetadataList = Reflect.getMetadata(decorators_1.metadataKeys.enums, target, propertyName);
            if (!enumsMetadataList?.length) {
                continue;
            }
            // Find first matching condition (first match wins)
            for (const enumsMetadata of enumsMetadataList) {
                if (applyEnumFiltering(property, enumsMetadata, evaluator, context)) {
                    break; // First match wins, stop processing
                }
            }
        }
    }
}
exports.Enums = Enums;
