// @sap/ux-specification/dist/specification/src/sync/common/decoration/steps/Message.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Message = void 0;
const ux_specification_types_1 = require("@sap/ux-specification-types");
const decorators_1 = require("../decorators");
const decorator_paths_1 = require("../decorator-paths");
const condition_evaluator_1 = require("../condition-evaluator");
const i18next_1 = __importDefault(require("i18next"));
/**
 * Adds a message to a schema element.
 *
 * @param element - The schema element to add the message to
 * @param root0 - Message parameters object
 * @param root0.text - The message text
 * @param root0.deletable - Whether the message is deletable (default: false)
 */
function addMessageToSchema(element, { text, deletable = false }) {
    if (!element[ux_specification_types_1.SchemaTag.messages]) {
        element[ux_specification_types_1.SchemaTag.messages] = [];
    }
    element[ux_specification_types_1.SchemaTag.messages].push({ text, deletable });
}
/**
 * Message step - adds messages to properties when conditions are met.
 * Multiple @message decorators can be applied to the same property.
 */
class Message {
    constructor() {
        this.name = 'message';
    }
    /**
     * Adds messages to properties whose conditions are met.
     *
     * @param target - The class prototype containing decorator metadata
     * @param definition - The schema definition to modify
     * @param context - Runtime context for condition evaluation
     */
    init(target, definition, context) {
        if (!definition?.properties || !target) {
            return;
        }
        const evaluator = new condition_evaluator_1.ConditionEvaluator(context);
        for (const propertyName in definition.properties) {
            const property = definition.properties[propertyName];
            const messageInfoList = Reflect.getMetadata(decorators_1.metadataKeys.message, target, propertyName);
            if (messageInfoList) {
                for (const messageInfo of messageInfoList) {
                    const { passed } = evaluator.evaluate(messageInfo);
                    if (passed && messageInfo.messageConfig) {
                        this.addConditionalMessage(messageInfo, property, evaluator);
                    }
                }
            }
        }
    }
    /**
     * Adds a translated message to the schema property if the metadata contains a MessageConfig.
     *
     * @param metadata - Decorator metadata containing the message config
     * @param definition - The schema property definition to add the message to
     * @param evaluator - Condition evaluator used to resolve PathNode params
     */
    addConditionalMessage(metadata, definition, evaluator) {
        // Handle MessageConfig format (from msg() helper)
        if (metadata.messageConfig && (0, decorator_paths_1.isMessageConfig)(metadata.messageConfig)) {
            const msgConfig = metadata.messageConfig;
            const resolvedParams = evaluator.resolveMessageParams(msgConfig.params);
            const messageText = i18next_1.default.t(msgConfig.i18nKey, resolvedParams);
            addMessageToSchema(definition, { text: messageText, deletable: msgConfig.deletable });
        }
    }
}
exports.Message = Message;
