// @sap/ux-specification/dist/specification/src/sync/v4/export/fpmCustomPage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.exportXMLFragmentConfig = exports.exportFPMCustomPageConfig = void 0;
exports.getExtensionExportResult = getExtensionExportResult;
const xml_js_1 = require("xml-js");
const ux_specification_types_1 = require("@sap/ux-specification-types");
const utils_1 = require("../utils/utils");
const macros_1 = require("../utils/macros");
const utils_2 = require("../../common/utils");
/**
 * Method compares values if values are different after converted to string.
 * It is used to determine if we need make change in XML file.
 *
 * @param {unknown} value1 First value to compare.
 * @param {unknown} value2 Second value to compare.
 * @returns {boolean} Returns true if values are same after values converted to string.
 */
function isExportValueEquals(value1, value2) {
    return value1 !== undefined && value2 !== undefined ? `${value1}` === `${value2}` : value1 === value2;
}
/**
 * Derives the namespace-prefixed name for a child aggregation element from its parent's name.
 *
 * @param {Element} parentElement Parent XML element whose prefix is inherited.
 * @param {string} localName Aggregation property name.
 * @returns {string} Prefixed element name (e.g. "macros:exportSettings").
 */
function getAggregationElementName(parentElement, localName) {
    const colonIdx = parentElement.name?.indexOf(':') ?? -1;
    return colonIdx > 0 ? `${parentElement.name.slice(0, colonIdx)}:${localName}` : localName;
}
/**
 * Checks whether a single aggregation wrapper already matches the config value.
 *
 * @param {Element} wrapperElement Aggregation wrapper XML element.
 * @param {unknown} configValue Config value to compare.
 * @returns {boolean} True if wrapper matches config (no re-export needed).
 */
function isSingleAggregationUnchanged(wrapperElement, configValue) {
    if (typeof configValue !== 'object' || configValue === null) {
        return false;
    }
    const controlElement = wrapperElement.elements?.[0];
    if (!controlElement) {
        return Object.keys(configValue).length === 0;
    }
    const controlClassName = controlElement.name?.split(':').pop() ?? '';
    const configObject = configValue[controlClassName];
    const expectedAttrs = configObject ?? configValue;
    const xmlAttrs = controlElement.attributes ?? {};
    const xmlKeys = Object.keys(xmlAttrs);
    const expectedKeys = Object.keys(expectedAttrs).filter((k) => typeof expectedAttrs[k] !== 'object');
    if (xmlKeys.length !== expectedKeys.length) {
        return false;
    }
    return xmlKeys.every((key) => `${xmlAttrs[key]}` === `${expectedAttrs[key]}`);
}
/**
 * Method deletes property by emoving attribute from element.
 * Method also considers if it is last attribute and element is complex property, then element would removed as well.
 *
 * @param {TraverseData} traverseData Schema and XML traverse data.
 * @param {string} name Name of property.
 */
function deleteProperty(traverseData, name) {
    const { element, schema, parent, contextPath } = traverseData;
    delete element.attributes[name];
    if (schema.metadata?.type === ux_specification_types_1.MacrosPropertyType.Property &&
        Object.keys(element.attributes).length === 0 &&
        parent?.element.elements) {
        // Last attribute is deleted for property - if element is property then remove property itself
        const context = contextPath[contextPath.length - 1];
        if (context) {
            parent.element.elements.splice(context.index, 1);
        }
    }
}
/**
 * Resolves a namespace-prefixed element name from a $ref and the root element's xmlns declarations.
 *
 * @param {string} name Element name to prefix.
 * @param {string} $ref Schema $ref (e.g., "#/definitions/sap.fe.macros.table.MassEdit").
 * @param {Element} rootElement Root XML element containing xmlns declarations.
 * @returns {string} Prefixed element name, or original name if no prefix resolved.
 */
function resolvePrefixedName(name, $ref, rootElement) {
    const definitionKey = (0, utils_2.getDefinitionKey)($ref);
    const lastDot = definitionKey.lastIndexOf('.');
    if (lastDot <= 0) {
        return name;
    }
    const namespace = definitionKey.slice(0, lastDot);
    const xmlnsAttrs = rootElement.attributes || {};
    // Exact match: xmlns:macrosTable="sap.fe.macros.table" → "macrosTable:name"
    for (const [attr, value] of Object.entries(xmlnsAttrs)) {
        if (value === namespace) {
            return attr === 'xmlns' ? name : `${attr.slice(6)}:${name}`;
        }
    }
    // Derive from closest parent: xmlns:macros="sap.fe.macros" → "macrosTable:name" for "sap.fe.macros.table"
    let closestPrefix = '';
    let closestNamespace = '';
    for (const [attr, value] of Object.entries(xmlnsAttrs)) {
        if (attr.startsWith('xmlns:') &&
            typeof value === 'string' &&
            namespace.startsWith(value + '.') &&
            value.length > closestNamespace.length) {
            closestNamespace = value;
            closestPrefix = attr.slice(6);
        }
    }
    if (closestNamespace) {
        const suffix = namespace.slice(closestNamespace.length + 1);
        const derivedPrefix = closestPrefix + suffix.charAt(0).toUpperCase() + suffix.slice(1);
        rootElement.attributes[`xmlns:${derivedPrefix}`] = namespace;
        return `${derivedPrefix}:${name}`;
    }
    return name;
}
/**
 * Callback method of traversing schema and XML.
 * Method exports value from page config object to XML element.
 *
 * @param {TraverseData} traverseData Schema and XML traverse data.
 * @param {SchemaDefinition} property Schema property to use for export.
 * @param {object} config Fragment of page config object.
 * @param {string} name Name of property to export.
 * @returns {boolean} Is new value differs.
 */
function detectElementDeletion(traverseData, property, config, name) {
    const { element, parentElement } = traverseData;
    if (element?.name === name) {
        if (traverseData.data?.nodesToDelete) {
            traverseData.data.nodesToDelete.push({
                element,
                parent: parentElement
            });
        }
        return true;
    }
    return false;
}
/**
 * Checks whether a single-aggregation export can be skipped because the wrapper is unchanged.
 *
 * @param {Element} element Parent XML element.
 * @param {string} name Property name.
 * @param {object} config Page config fragment.
 * @returns {boolean} True if export should be skipped.
 */
function isSingleAggregationExportSkipped(element, name, config) {
    const wrapper = element.elements?.find((el) => el.name === getAggregationElementName(element, name));
    return !!wrapper && isSingleAggregationUnchanged(wrapper, config[name]);
}
/**
 * Callback method of traversing schema and XML.
 * Method exports value from page config object to XML element.
 *
 * @param {TraverseData} traverseData Schema and XML traverse data.
 * @param {SchemaDefinition} property Schema property to use for export.
 * @param {object} config Fragment of page config object.
 * @param {string} name Name of property to export.
 * @returns {boolean} Is new value differs.
 */
function exportProperty(traverseData, property, config, name) {
    if (property.hidden) {
        return false;
    }
    const { element } = traverseData;
    if (name in config && config[name] !== undefined) {
        if (property.isViewNode) {
            return exportViewNode(traverseData, property, config, name);
        }
        if (element && !isExportValueEquals(element.attributes?.[name], config[name])) {
            if (property.metadata?.aggregationCardinality === ux_specification_types_1.MacrosAggregatioCardinality.Single &&
                isSingleAggregationExportSkipped(element, name, config)) {
                return false;
            }
            exportNonViewAttribute(element, name, config, property, traverseData);
            return true;
        }
    }
    else if (property.isViewNode && property.metadata?.overrideClass) {
        const overrideClass = property.metadata.overrideClass;
        const shortName = overrideClass.slice(overrideClass.lastIndexOf('.') + 1);
        if ((0, macros_1.findChildElementByLocalNameAndKey)(element, shortName, name)) {
            return exportViewNode(traverseData, property, config, name);
        }
    }
    else if (element?.attributes?.[name]) {
        deleteProperty(traverseData, name);
        return true;
    }
    else if (element?.elements && property.metadata?.aggregationCardinality === ux_specification_types_1.MacrosAggregatioCardinality.Single) {
        const idx = element.elements.findIndex((el) => el.name === getAggregationElementName(element, name));
        if (idx !== -1) {
            element.elements.splice(idx, 1);
            return true;
        }
    }
    return false;
}
/**
 * Dispatches export of an `isViewNode` property to override or standard handler.
 *
 * @param traverseData Schema and XML traverse context.
 * @param property Schema definition of the property being exported.
 * @param config Page config fragment.
 * @param name Property name.
 * @returns Whether the config value was written to XML.
 */
function exportViewNode(traverseData, property, config, name) {
    if (!traverseData.element.elements) {
        traverseData.element.elements = [];
    }
    return property.metadata?.overrideClass && traverseData.data?.rootElement
        ? exportOverrideNode(traverseData, property, config, name)
        : exportStandardViewNode(traverseData, property, config, name);
}
/**
 * Exports an annotation-based override node by finding or creating a typed child element,
 * then recursively traversing its schema to write override attributes.
 *
 * @param traverseData Schema and XML traverse context.
 * @param property Schema definition with overrideClass metadata.
 * @param config Page config fragment containing override values.
 * @param name Override key (used as the `key` attribute and config lookup).
 * @returns Whether any attributes were written.
 */
function exportOverrideNode(traverseData, property, config, name) {
    const { element, parentElement, fullSchema, contextPath } = traverseData;
    const overrideClassName = property.metadata.overrideClass;
    const shortName = overrideClassName.slice(overrideClassName.lastIndexOf('.') + 1);
    const prefixedName = resolvePrefixedName(shortName, `#/definitions/${overrideClassName}`, traverseData.data.rootElement);
    // Import tolerates an override placed at the parent control level (sibling of the aggregation
    // wrapper). Drop any such stray so this export doesn't leave a duplicate alongside the entry
    // inside the wrapper — its attributes are already in config from import.
    if (parentElement && parentElement !== element && parentElement.elements) {
        const strayIdx = parentElement.elements.findIndex((el) => el.name?.split(':').pop() === shortName && el.attributes?.key === name);
        if (strayIdx !== -1) {
            parentElement.elements.splice(strayIdx, 1);
        }
    }
    let overrideElement = (0, macros_1.findChildElementByLocalNameAndKey)(element, shortName, name);
    if (!overrideElement) {
        overrideElement = { name: prefixedName, type: 'element', attributes: {} };
        element.elements.push(overrideElement);
    }
    // 'key' identifies the annotation column — always set, not part of schema traversal
    overrideElement.attributes = overrideElement.attributes ?? {};
    overrideElement.attributes.key = name;
    const result = (0, macros_1.traverseSchema)({
        fullSchema,
        schema: property,
        element: overrideElement,
        contextPath: [...contextPath, { name, index: 0 }],
        config: config[name] ?? {},
        parent: traverseData,
        parentElement: element,
        data: traverseData.data
    }, exportProperty, detectElementDeletion);
    // Empty override (only 'key' attr, no child elements) — remove entirely
    const remainingAttrs = Object.keys(overrideElement.attributes ?? {}).filter((k) => k !== 'key');
    if (remainingAttrs.length === 0 && !overrideElement.elements?.length && element.elements) {
        const idx = element.elements.indexOf(overrideElement);
        if (idx !== -1) {
            element.elements.splice(idx, 1);
        }
    }
    return result;
}
/**
 * Exports a standard view node by creating a child element and recursively traversing.
 *
 * @param traverseData Schema and XML traverse context.
 * @param property Schema definition of the view node.
 * @param config Page config fragment.
 * @param name Property name used as the XML element name.
 * @returns Whether any nested properties were written.
 */
function exportStandardViewNode(traverseData, property, config, name) {
    const { element, fullSchema, contextPath } = traverseData;
    const targetElement = { name, type: 'element' };
    element.elements.push(targetElement);
    return (0, macros_1.traverseSchema)({
        fullSchema,
        schema: property,
        element: targetElement,
        contextPath: [...contextPath, { name, index: 0 }],
        config: config[name],
        parent: traverseData,
        parentElement: element,
        data: traverseData.data
    }, exportProperty, detectElementDeletion);
}
/**
 * Exports a non-view-node property value to XML as attribute, child element, or aggregation.
 *
 * @param element Current XML element to write into.
 * @param name Property name.
 * @param config Page config fragment.
 * @param property Schema definition.
 * @param traverseData Full traverse context for namespace resolution.
 */
function exportNonViewAttribute(element, name, config, property, traverseData) {
    const value = config[name];
    if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
        element.elements = element.elements || [];
        const additionalProperties = typeof property.additionalProperties === 'object' ? property.additionalProperties : {};
        const isSingleAggregation = property.metadata?.aggregationCardinality === ux_specification_types_1.MacrosAggregatioCardinality.Single;
        if (isSingleAggregation && '$ref' in additionalProperties && traverseData.data?.rootElement) {
            exportSingleAggregation(element, name, value, additionalProperties, traverseData.data.rootElement);
        }
        else {
            element.elements.push({ name, type: 'element', attributes: value });
        }
    }
    else {
        element.attributes = element.attributes || {};
        const attrValue = Array.isArray(value) ? value.join(',') : value;
        if (attrValue === '' || attrValue === undefined) {
            delete element.attributes[name];
        }
        else {
            element.attributes[name] = attrValue;
        }
    }
}
/**
 * Writes a single-cardinality aggregation as a wrapper + inner control element.
 *
 * @param element Parent XML element.
 * @param name Aggregation property name.
 * @param value Config object with attribute values.
 * @param additionalProperties Schema additionalProperties with $ref.
 * @param additionalProperties.$ref Control class definition reference.
 * @param rootElement Root XML element for namespace resolution.
 */
function exportSingleAggregation(element, name, value, additionalProperties, rootElement) {
    // Produces:  <macros:exportSettings>
    //              <macrosTable:ColumnExportSettings label="..." />
    //            </macros:exportSettings>
    const aggregationRef = additionalProperties.$ref;
    const aggregationName = getAggregationElementName(element, name);
    const definitionKey = (0, utils_2.getDefinitionKey)(aggregationRef);
    const className = definitionKey.slice(definitionKey.lastIndexOf('.') + 1);
    const controlName = resolvePrefixedName(className, aggregationRef, rootElement);
    // Config can be { ClassName: { attrs } } (nested from import) or flat { attrs }
    const innerConfig = value[className] ?? value;
    const controlAttrs = {};
    for (const [key, val] of Object.entries(innerConfig)) {
        if (typeof val !== 'object' || val === null) {
            controlAttrs[key] = val;
        }
    }
    const newWrapper = {
        name: aggregationName,
        type: 'element',
        elements: [{ name: controlName, type: 'element', attributes: controlAttrs }]
    };
    const existingIdx = element.elements.findIndex((el) => el.name === aggregationName);
    if (existingIdx !== -1) {
        element.elements[existingIdx] = newWrapper;
    }
    else {
        element.elements.push(newWrapper);
    }
}
/**
 * Deletes elements from their parent containers.
 *
 * @param nodesToDelete Array of elements marked for deletion with their parent references
 */
function deleteElements(nodesToDelete) {
    for (const deleteNode of nodesToDelete) {
        const { parent, element } = deleteNode;
        if (!parent?.elements) {
            continue;
        }
        const index = parent.elements.findIndex((el) => el === element);
        if (index !== -1) {
            parent.elements.splice(index, 1);
        }
    }
}
/**
 * Populates export results with data from extension.
 *
 * @param {ExtensionFileData} extensionData Extension data from parsed XML file.
 * @param {ExportResults} exportResults Export results API parameters needed for the export.
 * @param {object} config Extension of page config object.
 * @param {SchemaDefinition} jsonSchema Extesion definition schema.
 * @param {string} extensionType Extension type - view or fragment.
 * @param {ExtensionLogger | undefined} logger Logger class for logging messages.
 */
function getExtensionExportResult(extensionData, exportResults, config, jsonSchema, extensionType, logger) {
    const element = (0, macros_1.parseXML)(extensionData.file, false, logger);
    if (element) {
        const coreElement = (0, macros_1.getRootElement)(element);
        const traversionData = {
            nodesToDelete: [],
            rootElement: coreElement
        };
        const differs = (0, macros_1.traverseSchema)({
            fullSchema: jsonSchema,
            schema: jsonSchema,
            element: coreElement,
            contextPath: [],
            config,
            data: traversionData
        }, exportProperty, detectElementDeletion);
        if (differs) {
            if (traversionData.nodesToDelete?.length) {
                deleteElements(traversionData.nodesToDelete);
            }
            const fileContent = (0, xml_js_1.js2xml)(element, { spaces: 4 });
            const extensionExportResult = {
                dataSourceUri: extensionData.file.dataSourceUri,
                fileContent,
                changeIndicator: ux_specification_types_1.ChangeIndicator.Updated
            };
            if (extensionType === 'view') {
                exportResults.views = [extensionExportResult];
            }
            else if (extensionType === 'fragment') {
                exportResults.fragments.push(extensionExportResult);
            }
        }
    }
}
/**
 * Run through the given FPM Custom Page config and return respective XML View result.
 * If no changes in config, then view is not returned.
 *
 * @param exportParameters All API parameters needed for the export.
 * @param defaultExportResult Default export result.
 * @returns Object comprising the updated xml view
 */
const exportFPMCustomPageConfig = (exportParameters, defaultExportResult) => {
    const { views, manifest, jsonSchema, page, logger } = exportParameters;
    const exportResults = { ...defaultExportResult, manifest };
    let viewData = undefined;
    if (jsonSchema && Object.keys(jsonSchema).length > 0) {
        viewData = (0, utils_1.getPageCustomViewFile)(views || [], page.name, manifest, logger);
    }
    if (viewData) {
        getExtensionExportResult(viewData, exportResults, page.config, jsonSchema, 'view', logger);
    }
    return exportResults;
};
exports.exportFPMCustomPageConfig = exportFPMCustomPageConfig;
/**
 * Runs through the given config and pushes changed fragments into export results.
 *
 * @param {TransferParameterTypeV4} transferParameters List of transfer parameters passed from API.
 * @param {string} fragmentName Full name of the fragment.
 * @param {object} config Current (sub)object of the configuration file.
 */
const exportXMLFragmentConfig = (transferParameters, fragmentName, config) => {
    const { appSchema, exportResults, fragments, pageNameArray, logger } = transferParameters;
    const { manifest } = exportResults;
    let fragmentData = undefined;
    const extensionType = 'fragment';
    if (appSchema && Object.keys(appSchema).length > 0) {
        const targetManifestPage = (0, utils_1.getManifestPage)(manifest, pageNameArray[0]);
        fragmentData = (0, utils_1.getPageCustomExtensionFile)(fragments, targetManifestPage, manifest, extensionType, fragmentName);
    }
    if (fragmentData) {
        const definitionKey = (0, macros_1.getCustomExtensionFragmentDefinitionKey)(fragmentName);
        const targetConfigSchema = appSchema['definitions'][definitionKey];
        const tempSchema = {
            ...targetConfigSchema,
            definitions: appSchema['definitions']
        };
        getExtensionExportResult(fragmentData, exportResults, config, tempSchema, extensionType, logger);
    }
};
exports.exportXMLFragmentConfig = exportXMLFragmentConfig;
