// @sap/ux-specification/dist/specification/src/sync/v4/generate/utils/inlineEdit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.adjustInlineEdit = adjustInlineEdit;
const ux_specification_types_1 = require("@sap/ux-specification-types");
const propertiesForEdit_1 = require("./propertiesForEdit");
/**
 * Adjusts the application schema to configure inline edit functionality for a Fiori Elements page.
 *
 * **Prerequisites:** Decorator-based visibility and message rules for inlineEdit are applied at the page level
 * via `new ListReport(settings).init()` in `ListReportPage.generate()`. This function only enriches the
 * inlineEdit sub-properties when the property is already visible.
 *
 * 1. Configures the UI provider to use a switch control for enabling/disabling inline edit.
 * 2. Populates the `enabledFields` and `disabledFields` schema properties with all available
 *    editable columns, navigation properties, and field groups.
 * 3. Configures `connectedFields` to allow selection of field groups and individual editable properties.
 *
 * @param tableSchemaAdjustmentParameters - Parameters containing the app schema, V4 page configuration, and entity type metadata.
 * @param facetDefinitionId - The definition ID in the schema that contains the column definitions for the table/facet.
 */
function adjustInlineEdit(tableSchemaAdjustmentParameters, facetDefinitionId) {
    const { context, entityType } = tableSchemaAdjustmentParameters;
    const appSchema = context.schema.get();
    if (!entityType) {
        return;
    }
    if (appSchema.properties.inlineEdit && !appSchema.properties.inlineEdit[ux_specification_types_1.SchemaTag.hidden]) {
        // Update schema to use switch provider instead of generic provider for UI
        appSchema.properties.inlineEdit[ux_specification_types_1.SchemaTag.provider] = {
            name: 'Switch',
            parameters: { label: 'Inline Edit' }
        };
        const inlineEditProperties = appSchema.definitions.InlineEdit.properties;
        const columnsDefinitions = appSchema.definitions[facetDefinitionId];
        applyInlineEditTreeSchema(inlineEditProperties, columnsDefinitions, entityType);
    }
}
/**
 * Applies the tree schema structure to an inline edit field definition.
 *
 * @param fieldDefinitions - The schema definition object to be modified. Expected to contain
 *        `enabledFields`, `disabledFields`, and `connectedFields` keys which will be updated
 *        with tree schema structures.
 * @param columnsDefinitions - Read-only schema definition containing resolved column information.
 *        Used as input to build the field lists but not modified by this function.
 * @param entityType - The entity type providing property metadata and UI annotations.
 */
function applyInlineEditTreeSchema(fieldDefinitions, columnsDefinitions, entityType) {
    const fieldGroupsList = (0, propertiesForEdit_1.buildFieldGroupsList)(columnsDefinitions, entityType);
    const singleFieldList = (0, propertiesForEdit_1.buildSingleFieldList)(columnsDefinitions, entityType);
    const treeItem = {
        type: 'string',
        uiType: 'tree',
        creation: 'FieldGroup',
        oneOf: [...fieldGroupsList, ...singleFieldList]
    };
    ['enabledFields', 'disabledFields', 'connectedFields'].forEach((key) => {
        const propertyDef = fieldDefinitions[key];
        if (key === 'connectedFields') {
            const items = {
                type: 'string',
                uiType: 'tree',
                creation: 'FieldGroup',
                restrictGroups: true,
                oneOf: [
                    {
                        type: 'string',
                        header: true,
                        const: 'Groups',
                        oneOf: fieldGroupsList.map((fieldGroup) => ({ ...fieldGroup, selectionMode: 'single' })),
                        uniqueItems: true
                    },
                    {
                        type: 'array',
                        header: true,
                        const: 'Individual items',
                        oneOf: singleFieldList,
                        uniqueItems: true
                    }
                ]
            };
            propertyDef.items = items;
            propertyDef.uniqueItems = true;
        }
        else {
            Object.assign(propertyDef.items, treeItem);
            propertyDef.uniqueItems = true;
        }
    });
}
