// @sap/ux-specification/dist/specification/src/sync/v4/generate/utils/propertiesForEdit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isPropertyEditable = isPropertyEditable;
exports.getPropertyDescription = getPropertyDescription;
exports.getFieldGroupsWithMembers = getFieldGroupsWithMembers;
exports.getNavigationPropertiesForEdit = getNavigationPropertiesForEdit;
exports.getEditableFieldsForEdit = getEditableFieldsForEdit;
exports.buildFieldGroupsList = buildFieldGroupsList;
exports.buildSingleFieldList = buildSingleFieldList;
const odata_annotation_core_1 = require("@sap-ux/odata-annotation-core");
const common_1 = require("../../../common");
const utils_1 = require("../../utils/utils");
const UI_ANNOTATION_KEY = 'UI';
/**
 * Determines whether an entity property is editable based on its type and OData annotations.
 *
 * @param property - The entity property to evaluate for edibility.
 * @param options - Optional filter criteria for annotation flags.
 * @returns `true` if the property can be edited, `false` otherwise.
 */
function isPropertyEditable(property, options = {}) {
    if (!property) {
        return false;
    }
    const { immutable = false, computed = false, hidden = false, isImageURL = false, typeFilterFn } = options;
    const matchesHidden = hidden === null || !!property.annotations.UI?.Hidden === hidden;
    const matchesIsImageURL = isImageURL === null || !!property.annotations.UI?.IsImageURL === isImageURL;
    const matchesImmutable = immutable === null || !!property.annotations.Core?.Immutable === immutable;
    const matchesComputed = computed === null || !!property.annotations.Core?.Computed === computed;
    let matchType = property.type !== odata_annotation_core_1.EdmType.Stream && !property.type.startsWith('Collection(');
    if (typeFilterFn) {
        matchType = typeFilterFn(property, matchType);
    }
    return matchType && matchesHidden && matchesIsImageURL && matchesImmutable && matchesComputed;
}
/**
 * Gets the description/label for an entity property.
 *
 * @param property - The entity property to get the description for.
 * @returns The property label or the property name as fallback.
 */
function getPropertyDescription(property) {
    const label = property?.annotations?.Common?.Label;
    if (typeof label === 'string') {
        return label;
    }
    return property?.name ?? '';
}
/**
 * Gets UI.FieldGroup annotations from an entity type and returns their member fields with descriptions.
 *
 * @param entityType - The entity type containing UI annotations with potential FieldGroup definitions.
 * @param editableFields - Set of editable field names to filter field group members against.
 * @param prefixWithAt - When `true` (default), field group names are prefixed with `@` (e.g. `@UI.FieldGroup#qualifier`).
 *   Pass `false` to omit the prefix (e.g. for contexts where the `@` is already implied).
 * @returns Sorted array of FieldGroupInfo containing field group metadata and member field details.
 */
function getFieldGroupsWithMembers(entityType, editableFields, prefixWithAt = true) {
    const uiAnnotations = entityType?.annotations?.[UI_ANNOTATION_KEY] ?? {};
    return Object.keys(uiAnnotations)
        .filter((key) => key.startsWith(`${common_1.DataType.FieldGroup}${utils_1.QUALIFIER_SEPARATOR}`))
        .map((key) => {
        const annotation = uiAnnotations[key];
        const fields = annotation.Data.filter((record) => record.$Type === "com.sap.vocabularies.UI.v1.DataField" /* UIAnnotationTypes.DataField */ &&
            record.Value?.path &&
            editableFields.has(record.Value.path)).map((record) => {
            const propertyPath = record.Value?.path;
            const property = entityType.entityProperties.find((p) => p.name === propertyPath);
            return { value: propertyPath, description: getPropertyDescription(property) };
        });
        const prefix = prefixWithAt ? '@' : '';
        return {
            name: `${prefix}${annotation.term}${utils_1.QUALIFIER_SEPARATOR}${annotation.qualifier}`,
            fields
        };
    })
        .filter((fieldGroup) => fieldGroup.fields.length > 0)
        .sort((a, b) => a.name.localeCompare(b.name));
}
/**
 * Extracts navigation property paths from column definitions that are eligible for editing.
 *
 * Filters the column definitions to find paths containing navigation segments (indicated by `/`),
 * then validates each path by:
 * 1. Traversing each navigation segment to ensure it represents a to-one relationship (not a collection).
 * 2. Verifying that the final property in the path is editable according to `isPropertyEditable`.
 *
 * This ensures only single-valued navigation paths to editable properties are included,
 * excluding one-to-many relationships which cannot be edited.
 *
 * @param columnsDefinitions - The schema definition containing column/property definitions for the table.
 * @param entityType - The root entity type used to resolve navigation property paths.
 * @param options - Optional editability filter options to control which properties are treated as editable.
 * @returns An alphabetically sorted array of valid navigation property paths (e.g., `NavProperty/FieldName`).
 */
function getNavigationPropertiesForEdit(columnsDefinitions, entityType, options) {
    if (typeof columnsDefinitions !== 'object' || !columnsDefinitions?.properties) {
        return [];
    }
    return (0, utils_1.getColumnsEnum)(columnsDefinitions.properties)
        .filter((name) => {
        if (!name.includes('/')) {
            return false;
        }
        const navPropertyNames = name.split('/');
        let currEntityType = entityType;
        for (let i = 0; i < navPropertyNames.length - 1; i++) {
            const navPropertyName = navPropertyNames[i];
            const navPropertyDefinition = currEntityType.navigationProperties.find((np) => np.name === navPropertyName);
            // ignore all 1 : n properties
            if (!navPropertyDefinition || navPropertyDefinition.isCollection) {
                return false;
            }
            currEntityType = navPropertyDefinition.targetType;
        }
        const finalPropertyName = navPropertyNames[navPropertyNames.length - 1];
        const finalProperty = currEntityType.entityProperties.find((ep) => ep.name === finalPropertyName);
        return isPropertyEditable(finalProperty, options);
    })
        .sort((a, b) => a.localeCompare(b));
}
/**
 * Gets all editable fields (entity properties and navigation properties) for edit.
 *
 * @param columnsDefinitions - The schema definition containing column definitions.
 * @param entityType - The entity type providing property metadata.
 * @param options - Optional editability filter options to control which properties are treated as editable.
 * @returns Sorted array of field names (entity properties followed by navigation property paths).
 */
function getEditableFieldsForEdit(columnsDefinitions, entityType, options) {
    const entityFields = entityType.entityProperties
        .filter((property) => isPropertyEditable(property, options))
        .map((p) => p.name)
        .sort((a, b) => a.localeCompare(b));
    const navProperties = getNavigationPropertiesForEdit(columnsDefinitions, entityType);
    return [...entityFields, ...navProperties];
}
/**
 * Builds the FieldGroup nodes for an editable tree schema.
 *
 * @param columnsDefinitions - The schema definition containing column definitions.
 * @param entityType - The entity type providing property metadata and UI annotations.
 * @param options - Optional editability filter options to control which properties are treated as editable.
 * @param selectionMode - Selection mode for list item.
 * @param prefixWithAt - When `true` (default), field group names are prefixed with `@` (e.g. `@UI.FieldGroup#qualifier`).
 *   Pass `false` to omit the prefix (e.g. for contexts where the `@` is already implied).
 * @returns Array of FieldGroupNode objects representing FieldGroup selections.
 */
function buildFieldGroupsList(columnsDefinitions, entityType, options, selectionMode = 'multi', prefixWithAt = true) {
    const singleFields = getEditableFieldsForEdit(columnsDefinitions, entityType, options);
    const fieldGroups = getFieldGroupsWithMembers(entityType, new Set(singleFields), prefixWithAt);
    // Build FieldGroup nodes with nested children
    const fieldGroupNodes = fieldGroups.map((fieldGroup) => ({
        const: fieldGroup.name,
        group: 'FieldGroup',
        selectionMode,
        oneOf: fieldGroup.fields.map((field) => ({
            const: field.value,
            description: field.description,
            disabled: true,
            selectionMode: 'single'
        }))
    }));
    return [...fieldGroupNodes];
}
/**
 * Builds the single field nodes for an editable tree schema.
 *
 * @param columnsDefinitions - The schema definition containing column definitions.
 * @param entityType - The entity type providing property metadata and UI annotations.
 * @param options - Optional editability filter options to control which properties are treated as editable.
 * @returns Array of FieldNode objects representing individual field selections.
 */
function buildSingleFieldList(columnsDefinitions, entityType, options) {
    const singleFields = getEditableFieldsForEdit(columnsDefinitions, entityType, options);
    // Build single field nodes with descriptions for the tree picker UI
    const fieldNodes = singleFields.map((fieldName) => {
        // Navigation property paths (e.g., "Customer/Name") contain "/"
        const property = fieldName.includes('/')
            ? undefined
            : entityType.entityProperties.find((p) => p.name === fieldName);
        return {
            // Actual field identifier used when selected
            const: fieldName,
            // Human-readable label: use @Common.Label annotation if available
            description: property ? getPropertyDescription(property) : fieldName,
            group: 'Field'
        };
    });
    return [...fieldNodes];
}
