// @sap/ux-specification/dist/specification/src/sync/v4/utils/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QUALIFIER_SEPARATOR = exports.FIELD_SEPARATOR = exports.DATA_FIELD_FOR_ACTION_GROUP = exports.MIN_UI5_VERSION_FOR_TABLE_TYPE_RESTRICTIONS = void 0;
exports.hasEmbeddedComponent = hasEmbeddedComponent;
exports.addExtensionToArray = addExtensionToArray;
exports.getColumnsEnum = getColumnsEnum;
exports.addEnumForProperties = addEnumForProperties;
exports.addCustomColumnDefinition = addCustomColumnDefinition;
exports.addActionRecordToSchema = addActionRecordToSchema;
exports.addGroupActionRecordToSchema = addGroupActionRecordToSchema;
exports.addLineItemActionDefinition = addLineItemActionDefinition;
exports.addLineItemRecordToSchema = addLineItemRecordToSchema;
exports.appendLineItemsToSchema = appendLineItemsToSchema;
exports.addLineItemsTypeToSchema = addLineItemsTypeToSchema;
exports.addFragmentEnumForAnchor = addFragmentEnumForAnchor;
exports.addEnumForActionAnchor = addEnumForActionAnchor;
exports.addEnumForEnableMassEdit = addEnumForEnableMassEdit;
exports.initializeCreationModeSchema = initializeCreationModeSchema;
exports.addOneOfForCreationFields = addOneOfForCreationFields;
exports.initializePersonalizationSchema = initializePersonalizationSchema;
exports.findPageV4 = findPageV4;
exports.alignSchemaWithTemplateType = alignSchemaWithTemplateType;
exports.addCustomActionDefinition = addCustomActionDefinition;
exports.addHeaderFooterCustomActionDefinition = addHeaderFooterCustomActionDefinition;
exports.parseDefinitionName = parseDefinitionName;
exports.getManifestPage = getManifestPage;
exports.getPageCustomExtensionFile = getPageCustomExtensionFile;
exports.getPageCustomViewFile = getPageCustomViewFile;
exports.addCustomSectionDefinition = addCustomSectionDefinition;
exports.addCustomFilterFieldDefinition = addCustomFilterFieldDefinition;
exports.importSingleActionSettingsOfReference = importSingleActionSettingsOfReference;
exports.importGroupActionSettingsOfReference = importGroupActionSettingsOfReference;
exports.importManifestBasedActionMenuSettingsOfReference = importManifestBasedActionMenuSettingsOfReference;
exports.importActionSettingsOfReference = importActionSettingsOfReference;
exports.addCustomFieldDefinition = addCustomFieldDefinition;
exports.addManifestPathsToProperties = addManifestPathsToProperties;
exports.getInsertableForEntitySet = getInsertableForEntitySet;
exports.getDefaultTableType = getDefaultTableType;
exports.getTableTypeOptions = getTableTypeOptions;
exports.ensureEntitySetWithCorrectType = ensureEntitySetWithCorrectType;
exports.getTableTypeOptionsByContext = getTableTypeOptionsByContext;
exports.getPageManifestPath = getPageManifestPath;
const utils_1 = require("../../common/utils");
const importProject_1 = require("../../common/importProject");
const StableIdHelper_1 = require("./StableIdHelper");
const common_1 = require("../../common");
const i18next_1 = __importDefault(require("i18next"));
const __1 = require("../../..");
const ux_specification_types_1 = require("@sap/ux-specification-types");
const path_1 = require("path");
const DataFieldStrategy_1 = require("../../common/DataFieldStrategy");
const utils_2 = require("../import/utils");
const src_1 = require("@sap/ux-specification-types/src");
const v4controls = __importStar(require("../export/controls"));
const pages = __importStar(require("../export/pages"));
const v4_1 = require("@sap/ux-specification-types/src/v4");
const project_access_1 = require("@sap-ux/project-access");
const application_1 = require("../application");
const propertiesForEdit_1 = require("../generate/utils/propertiesForEdit");
const utils_3 = require("../sync-rules/utils");
/**
 * Map of table type options based on the detected default table type.
 * Each table type has a predefined set of compatible options and a default type.
 */
const TABLE_TYPE_OPTIONS_MAP = {
    [v4_1.TableTypeV4.AnalyticalTable]: {
        options: [v4_1.TableTypeV4.AnalyticalTable, v4_1.TableTypeV4.ResponsiveTable, v4_1.TableTypeV4.GridTable],
        defaultType: v4_1.TableTypeV4.AnalyticalTable
    },
    [v4_1.TableTypeV4.TreeTable]: {
        options: [v4_1.TableTypeV4.TreeTable, v4_1.TableTypeV4.ResponsiveTable, v4_1.TableTypeV4.GridTable],
        defaultType: v4_1.TableTypeV4.TreeTable
    },
    [v4_1.TableTypeV4.ResponsiveTable]: {
        options: [v4_1.TableTypeV4.ResponsiveTable, v4_1.TableTypeV4.GridTable],
        defaultType: v4_1.TableTypeV4.ResponsiveTable
    },
    [v4_1.TableTypeV4.GridTable]: {
        options: [v4_1.TableTypeV4.ResponsiveTable, v4_1.TableTypeV4.GridTable],
        defaultType: v4_1.TableTypeV4.ResponsiveTable
    }
};
/**
 * Default table type used as a fallback when entity set is undefined or capabilities cannot be determined.
 * ResponsiveTable is chosen as the safest and most universally applicable option.
 */
const FALLBACK_TABLE_TYPE = v4_1.TableTypeV4.ResponsiveTable;
/**
 * Fallback table type options used when entity set capabilities cannot be determined.
 * Derives from FALLBACK_TABLE_TYPE to ensure consistency across fallback scenarios.
 */
const FALLBACK_TABLE_TYPE_OPTIONS = TABLE_TYPE_OPTIONS_MAP[FALLBACK_TABLE_TYPE];
/**
 * Minimum UI5 version that supports capability-based table type restrictions.
 *
 * Before 1.142.0:
 * - Table type was optional
 * - All table types were available regardless of service capabilities
 * - No warning messages for missing table type or type-dependent properties
 *
 * From 1.142.0 onwards:
 * - Table type options are restricted based on entity set capabilities (analytical/hierarchical/standard)
 * - Warning messages are shown when table type is missing but type-dependent properties exist
 * - Schema enum values are dynamically restricted to match detected capabilities
 */
exports.MIN_UI5_VERSION_FOR_TABLE_TYPE_RESTRICTIONS = '1.142.0';
/**
 * Returns true when the given manifest section object contains an `embeddedComponent` property,
 * i.e. it is an HDM-style reuse component section rather than an XML-fragment custom section.
 *
 * @param section - The manifest section object to test.
 * @returns `true` if the section has an `embeddedComponent` property.
 */
function hasEmbeddedComponent(section) {
    return typeof section === 'object' && section !== null && ux_specification_types_1.TemplatePropertyName.EmbeddedComponent in section;
}
/**
 * Table type options for UI5 versions before 1.142.0, where capability-based restrictions
 * are not supported. All table types are available, with ResponsiveTable as the default.
 */
const LEGACY_VERSION_TABLE_TYPE_OPTIONS = {
    options: [v4_1.TableTypeV4.ResponsiveTable, v4_1.TableTypeV4.GridTable, v4_1.TableTypeV4.AnalyticalTable, v4_1.TableTypeV4.TreeTable],
    defaultType: v4_1.TableTypeV4.ResponsiveTable
};
exports.DATA_FIELD_FOR_ACTION_GROUP = 'DataFieldForActionGroup';
const UI_ANNOTATION_KEY = 'UI';
exports.FIELD_SEPARATOR = '::';
exports.QUALIFIER_SEPARATOR = '#';
/**
 * List of annotation-based action prefixes that should be included in the schema
 * even if they lack custom action properties (press, menu), so they can be validated
 * and removed during export if they don't match the config.
 */
const ANNOTATION_BASED_ACTION_PREFIXES = [`${ux_specification_types_1.DataFieldActionType.DataFieldForIntentBasedNavigation}${exports.FIELD_SEPARATOR}`];
const ACTION_TYPE = 'actionType';
const strategyTypeMap = new Map([
    ['', "com.sap.vocabularies.UI.v1.DataFieldForAction" /* UIAnnotationTypes.DataFieldForAction */],
    ['DataFieldForAction', "com.sap.vocabularies.UI.v1.DataFieldForAction" /* UIAnnotationTypes.DataFieldForAction */],
    ['DataFieldForActionGroup', "com.sap.vocabularies.UI.v1.DataFieldForActionGroup" /* UIAnnotationTypes.DataFieldForActionGroup */],
    ['CustomMenu', ux_specification_types_1.CustomUIAnnotationTypes.ManifestBasedActionMenu]
]);
const OP_ACTION_MENU_DEFINITION = {
    actionMenuDefinitionName: ux_specification_types_1.DefinitionName.ObjectPageCustomActionMenu,
    actionMenuActionsDefinitionName: ux_specification_types_1.DefinitionName.ObjectPageCustomActionMenuActions,
    actionName: ux_specification_types_1.DefinitionName.ObjectPageToolBarAction
};
const OP_HEADER_ACTION_MENU_DEFINITION = {
    actionMenuDefinitionName: ux_specification_types_1.DefinitionName.ObjectPageHeaderCustomActionMenu,
    actionMenuActionsDefinitionName: ux_specification_types_1.DefinitionName.ObjectPageHeaderCustomActionMenuActions,
    actionName: ux_specification_types_1.DefinitionName.ObjectPageHeaderAction
};
const LR_ACTION_MENU_DEFINITION = {
    actionMenuDefinitionName: ux_specification_types_1.DefinitionName.CustomActionMenu,
    actionMenuActionsDefinitionName: ux_specification_types_1.DefinitionName.CustomActionMenuActions,
    actionName: ux_specification_types_1.DefinitionName.ToolBarAction
};
const LR_ACTION_VIEW_MENU_DEFINITION = {
    actionMenuDefinitionName: ux_specification_types_1.DefinitionName.ViewCustomActionMenu,
    actionMenuActionsDefinitionName: ux_specification_types_1.DefinitionName.ViewCustomActionMenuActions,
    actionName: ux_specification_types_1.DefinitionName.ViewToolBarAction
};
const actionMenuDefinitionNameMap = new Map([
    [ux_specification_types_1.DefinitionName.CustomTableActionOP, OP_ACTION_MENU_DEFINITION],
    [ux_specification_types_1.DefinitionName.CustomFormActionOP, OP_ACTION_MENU_DEFINITION],
    [ux_specification_types_1.DefinitionName.CustomTableAction, LR_ACTION_MENU_DEFINITION],
    [ux_specification_types_1.DefinitionName.CustomHeaderActionOP, OP_HEADER_ACTION_MENU_DEFINITION],
    [ux_specification_types_1.DefinitionName.ViewCustomAction, LR_ACTION_VIEW_MENU_DEFINITION]
]);
/**
 * Returns a comparator function to sort objects by their property index.
 *
 * @returns A comparator function that compares two objects based on their property index.
 */
function sortByPropertyIndex() {
    return ([_key1, value1], [_key2, value2]) => value1[ux_specification_types_1.SchemaTag.propertyIndex] - value2[ux_specification_types_1.SchemaTag.propertyIndex];
}
/**
 * Sort function for properties in schema definitions.
 *
 * @param definition - definition which contains list of properties, unsorted
 * @returns - list of properties, sorted by property index
 */
function sortPropertiesRecords(definition) {
    definition = typeof definition === 'object' ? definition : {};
    const unsortedObjArr = [...Object.entries(definition.properties)];
    if (!definition.properties || Object.keys(definition.properties).length === 0) {
        return unsortedObjArr;
    }
    return unsortedObjArr.sort(sortByPropertyIndex());
}
/**
 * Method receives extension object and returns title/text.
 *
 * @param extension - custom extension object from manifest.
 * @returns Title/text of custom extension.
 */
function getExtensionText(extension) {
    let text;
    if (typeof extension === 'object') {
        if ('header' in extension) {
            text = extension.header;
        }
        else if ('text' in extension) {
            text = extension.text;
        }
        else if ('title' in extension) {
            text = extension.title;
        }
        else if ('label' in extension) {
            text = extension.label;
        }
    }
    else if (typeof extension === 'string') {
        text = extension;
    }
    return text;
}
/**
 * Adds a custom extension to the right position of a sorted array.
 *
 * @param extension - custom extension
 * @param sortedExtensionsArray - array of all entries(annotation entries and extensions), to be enhanced
 * @param targetIndex - index of the anchor entry
 * @param extId - ID of the new custom extension
 * @param newExtensionReference - schema reference of the new extension
 * @param newExtensionReference.$ref - The reference to the schema definition for the new extension.
 * @param newExtensionReference.description - A description of the new extension.
 */
function addExtensionToSortedArray(extension, sortedExtensionsArray, targetIndex, extId, newExtensionReference) {
    const placement = extension.position?.placement || ux_specification_types_1.v4.Placement.After;
    if (placement === ux_specification_types_1.v4.Placement.Before) {
        sortedExtensionsArray.splice(targetIndex, 0, [extId, newExtensionReference]);
    }
    else {
        sortedExtensionsArray.splice(targetIndex + 1, 0, [extId, newExtensionReference]);
    }
}
/**
 * Adds remaining custom extensions that might self-reference and were not added in previous steps.
 *
 * @param {CustomExtensions} remainingExtensions - list of custom extensions that are left for post-processing
 * @param {string} extId - current id of remaining extensions
 * @param {CustomExtension} extension - current extension of remaining custom extensions
 * @param {[string, unknown]} sortedExtArray - sorted array of annotation-based extensions, to be updated
 * @param newCustomExt - description based on extension in manifest
 * @param newCustomExt.$ref - The reference to the schema definition for the new extension.
 * @param newCustomExt.description - A description of the new custom extension.
 * @param newCustomExt.keys - An array of key objects, each containing a name and value for the custom extension.
 */
function addRemainingTargetByBefore(remainingExtensions, extId, extension, sortedExtArray, newCustomExt) {
    const targetRemainingIndex = Object.entries(remainingExtensions)
        .filter(([key]) => key !== extId)
        .findIndex(([_key, entry]) => entry.position?.placement === 'Before' && entry.position?.anchor === extension.position?.anchor);
    if (targetRemainingIndex === -1) {
        sortedExtArray.splice(sortedExtArray.length, 0, [extId, newCustomExt]);
        delete remainingExtensions[extId];
    }
}
/**
 * Returns custom extension details like $ref, description, keys and action type.
 *
 * @param extension - custom extensions config
 * @param extensionDefinitionName - custom extension definition name, different in LR and OP
 * @param extId - custom extension id
 * @param actionType - action type
 * @returns Custom extension details
 */
function getExtensionDetails(extension, extensionDefinitionName, extId, actionType) {
    const refDefinition = extensionDefinitionName.includes(ux_specification_types_1.DefinitionName.ObjectPageCustomSectionFragment)
        ? `${extensionDefinitionName}<${extId}>`
        : extensionDefinitionName;
    const description = getExtensionText(extension);
    return {
        $ref: `${utils_1.DEFINITION_LINK_PREFIX}${refDefinition}`,
        description,
        keys: [{ name: ux_specification_types_1.SchemaKeyName.key, value: `${extId}` }],
        ...(actionType !== undefined && { actionType })
    };
}
/**
 * Post-processing of custom extensions that are left after the first loop of adding extensions, i.e. custom extensions referring to other custom extensions.
 *
 * @param {CustomExtensions} remainingExtensions - An object containing the remaining extensions to process.
 * @param {string} extensionDefinitionName - The name of the extension definition to refer to.
 * @param {[string, unknown][]} sortedExtArray - A sorted array of existing extensions represented as key-value pairs.
 * @param {ExtensionLogger} logger - The logger used for logging errors and warnings during the adjustment process.
 * @param {boolean} [forceAdd] - Flag indicating whether to force adding extensions even if proper anchor/placement is not found.
 */
function adjustCustomExtReferringToCustomExt(remainingExtensions, extensionDefinitionName, sortedExtArray, logger, forceAdd = false) {
    const initialRemaining = Object.entries(remainingExtensions).length;
    for (const extId of Object.keys(remainingExtensions)) {
        // make sure that we do not add the same extension twice
        const isAlreadyAdded = sortedExtArray.find((element) => element[0] === extId) !== undefined;
        if (isAlreadyAdded) {
            delete remainingExtensions[extId];
            continue;
        }
        const extension = remainingExtensions[extId];
        // Detect description based on extension in manifest
        const newCustomExt = getExtensionDetails(extension, extensionDefinitionName, extId, ux_specification_types_1.ActionType.Custom);
        if (extension.position?.anchor && extension.position?.placement) {
            const targetIndex = sortedExtArray.findIndex((element) => element[0] === extension.position.anchor);
            if (targetIndex > -1) {
                addExtensionToSortedArray(extension, sortedExtArray, targetIndex, extId, newCustomExt);
                delete remainingExtensions[extId];
            }
            else if (extId === extension.position.anchor ||
                Object.keys(remainingExtensions).indexOf(extension.position.anchor) === -1) {
                (0, __1.log)(logger, {
                    severity: "error" /* LogSeverity.Error */,
                    message: i18next_1.default.t('INVALIDANCHOR', { id: extId, anchor: extension.position.anchor }),
                    location: {
                        path: ux_specification_types_1.MANIFESTPATH,
                        range: [ux_specification_types_1.ManifestSection.ui5]
                    }
                });
                sortedExtArray.splice(sortedExtArray.length, 0, [extId, newCustomExt]);
                delete remainingExtensions[extId];
            }
            else if (forceAdd) {
                addRemainingTargetByBefore(remainingExtensions, extId, extension, sortedExtArray, newCustomExt);
            }
        }
        else if (forceAdd) {
            addExtensionToSortedArray(extension, sortedExtArray, sortedExtArray.length, extId, newCustomExt);
            delete remainingExtensions[extId];
        }
    }
    const endRemaining = Object.entries(remainingExtensions).length;
    if (endRemaining > 0 && initialRemaining !== endRemaining) {
        adjustCustomExtReferringToCustomExt(remainingExtensions, extensionDefinitionName, sortedExtArray, logger);
    }
    else if (endRemaining > 0) {
        adjustCustomExtReferringToCustomExt(remainingExtensions, extensionDefinitionName, sortedExtArray, logger, true);
    }
}
/**
 * Returns anchor for extension. Priority:
 * 1. 'anchor' property in extension;
 * 2. When 'anchor' is not defined in extension and does calculation based placement and current list of extension and annotation entries.
 *
 * @param extension - extension in manifest
 * @param sortedExtensionsArray - sorted array of annotation-based nodes, to be enhanced by extensions
 * @returns Anchor for extension
 */
function getExtensionAnchor(extension, sortedExtensionsArray) {
    // Default placement is After
    const placement = extension.position?.placement || ux_specification_types_1.v4.Placement.After;
    let anchor = extension.position?.anchor;
    if (!anchor) {
        // Simulate runtime behavior - if there no anchor, then logic uses first or last annotation node
        // In some cases last extension can be previously inserted extension - in such case we still need use annotation node
        const index = placement === ux_specification_types_1.v4.Placement.Before ? 0 : sortedExtensionsArray.length - 1;
        anchor = sortedExtensionsArray[index] ? sortedExtensionsArray[index][0] : undefined;
    }
    return anchor;
}
/**
 * Preprocesses and returns list of local extension ids that needs to be skipped from addition.
 *
 * @param {CustomExtensions} localExtensions - list of extensions in manifest
 * @param {[string, unknown][]} annotationEntries - sorted array of annotation-based nodes, to be enhanced by extensions
 * @returns List of local extensions that has no annotation based anchor
 */
function getLocalExtensionsToSkip(localExtensions, annotationEntries) {
    // collect local extensions that has no annotation based anchor
    const ignoredIds = [];
    for (const extensionId in localExtensions) {
        const anchor = localExtensions[extensionId].position?.anchor;
        const hasAnnotationAnchor = annotationEntries.some((annotationEntry) => annotationEntry[0] === anchor);
        if (anchor && !hasAnnotationAnchor) {
            ignoredIds.push(extensionId);
        }
    }
    return ignoredIds;
}
/**
 * Method returns position of passed extension using calculation for merged annotation nodes approach.
 * Currently is used for custom subsections - annotation merges subsection if there is no any child collecation facet.
 *
 * @param extension - custom extension
 * @param extensions - list of extensions in manifest
 * @param sortedExtensionsArray - array of all entries(annotation entries and extensions), to be enhanced
 * @returns Index of extension using merged annotation nodes approach.
 */
function getExtensionIndexByMergedApproach(extension, extensions, sortedExtensionsArray) {
    const placement = extension.position?.placement || ux_specification_types_1.v4.Placement.After;
    const originalAnchor = extension.position?.anchor;
    if (originalAnchor) {
        // Custom extension is anchored to any extension
        return sortedExtensionsArray.length;
    }
    // Custom extension is not anchored to any extension
    // Inner method to find next available index before or after annotation node
    const findNextIndex = (index, condition) => {
        for (const sortExtension of sortedExtensionsArray) {
            const innerExtension = extensions[sortExtension[0]];
            if (condition(innerExtension)) {
                // Condition is fulfilled - no need to increase index
                break;
            }
            // Increase index and check next ordered extension
            index++;
        }
        return index;
    };
    let index = 0;
    if (placement === ux_specification_types_1.v4.Placement.Before) {
        // Special case when no anchor but placement is "Before" - then rendered as very first
        // Find index before first annotation node
        index = findNextIndex(0, (innerExtension) => !innerExtension);
    }
    else {
        // Special case when no anchor but placement is "After" - then rendered right after last annotation node
        index = findNextIndex(-1, (innerExtension) => !!(innerExtension &&
            // Ignore extension without anchor but with position "Before"
            (innerExtension?.position?.anchor || innerExtension?.position?.placement !== ux_specification_types_1.v4.Placement.Before)));
    }
    return index > -1 ? index : 0;
}
/**
 *
 * @param extension - custom extension
 * @param extensions - list of extensions in manifest
 * @param sortedExtensionsArray - array of all entries(annotation entries and extensions), to be enhanced
 * @param anchor - target anchor to check
 * @param ignoreAnnotationAnchors - calculation should ignore anchors referenced to annotation nodes. Used when annotations sections are merged
 * @returns Index of extension using merged annotation nodes approach.
 */
function getExtensionIndex(extension, extensions, sortedExtensionsArray, anchor, ignoreAnnotationAnchors = false) {
    return !ignoreAnnotationAnchors
        ? sortedExtensionsArray.findIndex((element) => element[0] === anchor)
        : getExtensionIndexByMergedApproach(extension, extensions, sortedExtensionsArray);
}
/**
 * Error handler for invalid extensions.
 *
 * @param logger - logger for error messages
 * @param extensionId - ID of the extension
 * @param sortedExtensionsArray - sorted array of annotation-based nodes, to be enhanced by extensions
 * @param newExtensionReference - schema reference that had been prepared before
 * @param newExtensionReference.$ref - The reference to the schema definition for the new extension.
 * @param newExtensionReference.description - A description of the new extension.
 * @param newExtensionReference.keys - An array of key objects, each containing a name and value for the custom extension.
 * @param newExtensionReference.actionType - The action type associated with the new extension.
 * @param localExtensions - list of extensions for processing
 */
function handleInvalidExtension(logger, extensionId, sortedExtensionsArray, newExtensionReference, localExtensions) {
    (0, __1.log)(logger, {
        severity: "error" /* LogSeverity.Error */,
        message: i18next_1.default.t('INVALIDEXTENSION', { id: extensionId }),
        location: {
            path: ux_specification_types_1.MANIFESTPATH
        }
    });
    sortedExtensionsArray.push([extensionId, newExtensionReference]);
    delete localExtensions[extensionId];
}
/**
 * Checks if an extension ID represents an annotation-based action.
 * Annotation-based actions should be included in the schema even if they lack custom action
 * properties (press, menu), so they can be validated and removed during export if they don't match the config.
 *
 * @param {string} extensionId - The extension ID to check.
 * @returns {boolean} True if the extension is an annotation-based action, false otherwise.
 */
function isAnnotationBasedAction(extensionId) {
    return ANNOTATION_BASED_ACTION_PREFIXES.some((prefix) => extensionId.startsWith(prefix));
}
/**
 * Handles orphaned annotation-based actions that exist in manifest but not in annotations.
 * These actions must be added to the schema to enable cleanup during export.
 *
 * @param {string} extensionId - The extension ID.
 * @param {any} extension - The extension object.
 * @param {string} extensionDefinitionName - The definition name associated with the extensions.
 * @param {Definition} [appSchema] - Optional application schema for creating specific definitions.
 * @param {ActionDetails[]} sortedExtensionsArray - The array to add the action to.
 */
function handleOrphanedAnnotationAction(extensionId, extension, extensionDefinitionName, appSchema, sortedExtensionsArray) {
    // This is an annotation-based action from manifest that doesn't exist in annotations.
    // IMPORTANT: We must add these orphaned actions to the schema to enable cleanup.
    //
    // When an action is renamed (e.g., IBN action's semantic object changes), the old
    // action key remains in the manifest but disappears from the config (annotations).
    // By adding it to the schema, the export process can detect this mismatch and remove
    // the orphaned entry from the manifest. If we don't add it to the schema, it will be
    // ignored and remain as garbage in the manifest forever.
    //
    // Process:
    // 1. Add orphaned action to schema with Standard action type
    // 2. During export: compare schema with config
    // 3. Action in schema but NOT in config → remove from manifest
    // 4. Action in schema AND in config → keep in manifest
    if (appSchema) {
        // Create specific definition name and schema like "ToolBarActionGroup<DataFieldForActionGroup::AnnoActionMenu>"
        const { actionName } = determineActionDefinitionNames(extensionDefinitionName);
        const preparedRef = (0, common_1.prepareRef)(extensionId);
        const specificDefinitionName = (0, common_1.getFacetDefinitionKey)(actionName, preparedRef);
        // Create the schema definition for this action
        (0, common_1.parseSchemaDefinition)(actionName, preparedRef, appSchema);
        const correctedExtensionReference = getExtensionDetails(extension, specificDefinitionName, extensionId, ux_specification_types_1.ActionType.Standard);
        sortedExtensionsArray.push([extensionId, correctedExtensionReference]);
    }
    else {
        // Fallback if no appSchema provided
        const correctedExtensionReference = getExtensionDetails(extension, extensionDefinitionName, extensionId, ux_specification_types_1.ActionType.Standard);
        sortedExtensionsArray.push([extensionId, correctedExtensionReference]);
    }
}
/**
 * Recursively checks if an action exists in annotation entries, including nested actions within action groups.
 *
 * @param {string} extensionId - The unique identifier for the extension to search for.
 * @param {ActionDetails[]} annotationEntries - Array of actions to search through.
 * @param {Definition | undefined} appSchema - Optional application schema to look up action group definitions.
 * @returns {boolean} True if the action exists at any level (top-level or nested in groups).
 */
function actionExistsInAnnotations(extensionId, annotationEntries, appSchema) {
    for (const [id, actionDef] of annotationEntries) {
        // Check top-level match
        if (id === extensionId) {
            return true;
        }
        // Check if this is an action group with nested actions
        const actionDefinition = actionDef;
        if (appSchema && actionDefinition.$ref) {
            const defKey = actionDefinition.$ref.replace('#/definitions/', '');
            const definition = appSchema.definitions?.[defKey];
            // If it's an action group, check its nested actions (properties)
            if (definition && definition.properties) {
                const nestedActions = Object.entries(definition.properties).map(([nestedId, nestedDef]) => [nestedId, nestedDef]);
                // Recursively search nested actions
                if (actionExistsInAnnotations(extensionId, nestedActions, appSchema)) {
                    return true;
                }
            }
        }
    }
    return false;
}
/**
 * Processes annotation-based actions with unsupported extension types.
 * Either skips duplicates or handles orphaned actions.
 *
 * @param {string} extensionId - The extension ID.
 * @param {any} extension - The extension object.
 * @param {string} extensionDefinitionName - The definition name.
 * @param {Definition} [appSchema] - Optional application schema.
 * @param {ActionDetails[]} annotationEntries - Array of annotation entries.
 * @param {ActionDetails[]} sortedExtensionsArray - The sorted extensions array.
 */
function processAnnotationBasedAction(extensionId, extension, extensionDefinitionName, appSchema, annotationEntries, sortedExtensionsArray) {
    // Check if this annotation-based action exists in annotationEntries (including nested in action groups)
    const existsInAnnotations = actionExistsInAnnotations(extensionId, annotationEntries, appSchema);
    if (existsInAnnotations) {
        // Action already exists from annotations - skip it to avoid duplicates
        return;
    }
    // Handle orphaned annotation-based action
    handleOrphanedAnnotationAction(extensionId, extension, extensionDefinitionName, appSchema, sortedExtensionsArray);
}
/**
 * Adds extension entries to a sorted array of extensions based on specific rules,
 * including using anchors, callbacks, and custom templates.
 *
 * @param {CustomExtensions} extensions - The object containing custom extensions to be added to the array.
 * @param {ActionDetails[]} sortedExtensionsArray - The array of sorted extensions to which new extensions will be added.
 * @param {string} extensionDefinitionName - The definition name associated with the extensions being added.
 * @param {TemplatePropertyName | TemplatePropertyName[]} [templatePropertyName] -
 * The property name(s) to check for existence in an extension; used to determine compatibility.
 * @param {ExtensionLogger} [logger] - An optional logger to handle logging of invalid extensions or other events.
 * @param {boolean} [ignoreAnnotationAnchors] - A flag to indicate whether extension anchors should be ignored.
 * @param {Definition} [appSchema] - Optional application schema for creating specific definitions for annotation-based actions.
 * @returns {ActionDetails[]} The updated sorted extensions array, including newly added and valid extensions.
 */
function addExtensionToArray(extensions, sortedExtensionsArray, extensionDefinitionName, templatePropertyName = ux_specification_types_1.TemplatePropertyName.Template, logger, ignoreAnnotationAnchors = false, appSchema) {
    templatePropertyName = Array.isArray(templatePropertyName) ? templatePropertyName : [templatePropertyName];
    const annotationEntries = [...sortedExtensionsArray];
    const localExtensions = { ...extensions };
    const localExtensionsToSkip = getLocalExtensionsToSkip(localExtensions, annotationEntries);
    if (!localExtensions) {
        return [];
    }
    for (const extensionId of Object.keys(localExtensions)) {
        const extension = localExtensions[extensionId];
        if (localExtensionsToSkip?.includes(extensionId)) {
            continue;
        }
        const newExtensionReference = getExtensionDetails(extension, extensionDefinitionName, extensionId, ux_specification_types_1.ActionType.Custom);
        if (typeof extension !== 'object') {
            handleInvalidExtension(logger, extensionId, sortedExtensionsArray, newExtensionReference, localExtensions);
            continue;
        }
        const hasUnsupportedExtensionType = templatePropertyName.every((propertyName) => !(propertyName in extension));
        if (hasUnsupportedExtensionType) {
            const isAnnotationBased = isAnnotationBasedAction(extensionId);
            if (isAnnotationBased) {
                processAnnotationBasedAction(extensionId, extension, extensionDefinitionName, appSchema, annotationEntries, sortedExtensionsArray);
                delete localExtensions[extensionId];
                continue;
            }
            else {
                // Skip regular entries (columns, actions, etc.) that don't have required properties
                delete localExtensions[extensionId];
                continue;
            }
        }
        const anchor = getExtensionAnchor(extension, annotationEntries);
        if (anchor) {
            const targetIndex = getExtensionIndex(extension, extensions, sortedExtensionsArray, anchor, ignoreAnnotationAnchors);
            if (targetIndex > -1) {
                addExtensionToSortedArray(extension, sortedExtensionsArray, targetIndex, extensionId, newExtensionReference);
                delete localExtensions[extensionId];
            }
        }
        else {
            // Empty array - no any anchor to reference to
            sortedExtensionsArray.push([extensionId, newExtensionReference]);
            delete localExtensions[extensionId];
        }
    }
    if (Object.keys(localExtensions).length > 0) {
        //Second loop is necessary for custom columns that refer to custom columns
        adjustCustomExtReferringToCustomExt(localExtensions, extensionDefinitionName, sortedExtensionsArray, logger, false);
    }
    return sortedExtensionsArray;
}
/**
 * Extracts the column name from the keys metadata of a column definition.
 * Returns the value from 'Target' or 'Value' key, with UI namespace stripped.
 *
 * @param keys - Array of schema definition keys from the column definition.
 * @returns The column name with proper separators, or undefined if not found.
 */
function getColumnNameFromKeys(keys) {
    if (!keys?.length) {
        return undefined;
    }
    const keyEntry = keys.find((k) => k.name === ux_specification_types_1.SchemaKeyName.target || k.name === ux_specification_types_1.SchemaKeyName.value);
    if (!keyEntry?.value) {
        return undefined;
    }
    // Strip UI. namespace prefix (e.g., "UI.FieldGroup#qual" -> "FieldGroup#qual")
    return keyEntry.value.replace(/\bUI\./g, '');
}
/**
 * Returns column names for enum, extracted from the keys metadata of each column.
 * Only columns with 'Target' or 'Value' keys are included (excludes actions).
 *
 * @param lineItems - Current lineItems definition, with all records.
 * @returns Column names with proper separators (/ for navigation, # for qualifiers).
 */
function getColumnsEnum(lineItems) {
    if (typeof lineItems === 'boolean') {
        return [];
    }
    const columnEnum = [];
    for (const definition of Object.values(lineItems)) {
        const columnName = getColumnNameFromKeys(definition?.[ux_specification_types_1.SchemaTag.keys]);
        if (columnName) {
            columnEnum.push(columnName);
        }
    }
    return columnEnum;
}
/**
 * Removes the field prefix from a given key based on the defined FIELD_SEPARATOR.
 *
 * @param {string} key - The string key that contains a prefix and a value separated by FIELD_SEPARATOR.
 * @returns {string} The key without its prefix, retaining the portion after the FIELD_SEPARATOR.
 */
function removeDataFieldPrefix(key) {
    const [_, ...value] = key.split(exports.FIELD_SEPARATOR);
    return value.join(exports.FIELD_SEPARATOR);
}
/**
 * Extracts and processes the columns defined in the given lineItems or Definition object,
 * returning an array of CreationFieldOneOfOption.
 *
 * @param {Definition} [columns] - An object containing column definitions where keys represent field identifiers
 *                                     and values are field metadata.
 * @returns {CreationFieldOneOfOption[]} An array of field options, each containing field-specific metadata.
 */
function getCreationFieldOptions(columns = {}) {
    const columnOneOfEnum = [];
    // Extracts fields from line item properties, excluding field groups (they are handled separately)
    for (const key of Object.keys(columns)) {
        const columnName = removeDataFieldPrefix(key);
        if (columnName && !columnName.includes(src_1.CreationFieldType.FieldGroup)) {
            columnOneOfEnum.push({
                const: columnName,
                description: columns[key].description,
                groupType: src_1.CreationFieldType.Field
            });
        }
    }
    return columnOneOfEnum;
}
/**
 * Extracts field group options from the annotations of the provided entity type.
 *
 * @param {EntityType} [entityType] The entity type which contains annotations from which field groups are derived.
 * @returns {CreationFieldOneOfOption[]} An array of field group options extracted from the entity type annotations.
 */
function getFieldGroupsFromEntityType(entityType) {
    const uiAnnotations = entityType?.annotations?.[UI_ANNOTATION_KEY] ?? {};
    const annotationKeys = Object.keys(uiAnnotations);
    const fieldGroupKeys = annotationKeys.filter((key) => key.startsWith(`${src_1.CreationFieldType.FieldGroup}${exports.QUALIFIER_SEPARATOR}`));
    return fieldGroupKeys.map((key) => {
        const { term, qualifier, Label: label } = uiAnnotations[key];
        const option = {
            const: `${term}${exports.FIELD_SEPARATOR}${qualifier}`,
            groupType: src_1.CreationFieldType.FieldGroup
        };
        if (label) {
            option.description = label;
        }
        return option;
    });
}
/**
 * Adds enum values to the properties definition of a custom column.
 *
 * @param lineItems - current lineItems definition, with all records.
 * @param customColumnDefinition - current custom column definition, in the app schema
 */
function addEnumForProperties(lineItems, customColumnDefinition) {
    customColumnDefinition.properties.properties.items['enum'] = getColumnsEnum(lineItems.properties);
}
/**
 * Adds the propertyIndex to each column and adds the enum values for properties of custom columns.
 *
 * @param sortedColumnsArray - sorted array of all columns
 * @param {Definition} appSchema - app specific JSON schema
 * @param lineItems - content of the line item annotation
 * @param {DefinitionName} customColumnDefinitionName - custom column definition name, distinguishes LR from OP
 */
function addColumnPropertyIndexAndEnum(sortedColumnsArray, appSchema, lineItems, customColumnDefinitionName = ux_specification_types_1.DefinitionName.CustomColumn) {
    if (customColumnDefinitionName === ux_specification_types_1.DefinitionName.CustomColumn ||
        customColumnDefinitionName.indexOf(ux_specification_types_1.DefinitionName.ViewCustomColumn) > -1) {
        //Table custom column: Add enum for properties
        const customColumnDefinition = appSchema.definitions[customColumnDefinitionName];
        addEnumForProperties(lineItems, customColumnDefinition);
    }
    //Add property index
    for (let index = 0; index < sortedColumnsArray.length; index++) {
        const column = sortedColumnsArray[index][1];
        column[ux_specification_types_1.SchemaTag.propertyIndex] = index;
    }
}
/**
 * Common function for enhancing the LineItems definition of app schema by custom columns.
 *
 * @param appSchema - app-specific JSON schema
 * @param v4Page - actual page in the manifest
 * @param logger - logger for error messages
 * @param customColumnDefinitionName - custom column definition name, distinguishes LR from OP
 * @param sectionId - identifier of the current object page section in schema
 * @param sectionIdInManifest - identifier of the current object page section in manifest
 */
function addCustomColumnDefinition(appSchema, v4Page, logger, customColumnDefinitionName = ux_specification_types_1.DefinitionName.CustomColumn, sectionId, sectionIdInManifest) {
    const lineItemId = sectionId || ux_specification_types_1.DefinitionName.LineItems;
    const lineItems = appSchema.definitions[lineItemId];
    if (!lineItems) {
        return;
    }
    //sort line items
    let sortedColumnsArray = sortPropertiesRecords(lineItems);
    const id = sectionIdInManifest ? sectionIdInManifest.replace(/::/g, '/') : `@${"com.sap.vocabularies.UI.v1.LineItem" /* UIAnnotationTerms.LineItem */}`;
    const columns = v4Page.options?.settings?.controlConfiguration?.[id]?.['columns'];
    if (columns) {
        sortedColumnsArray = addExtensionToArray(columns, sortedColumnsArray, customColumnDefinitionName, undefined, logger);
    }
    //Adjust propertyIndex and add properties' enum
    addColumnPropertyIndexAndEnum(sortedColumnsArray, appSchema, lineItems, customColumnDefinitionName);
    const sortedColumnsAsObject = {};
    sortedColumnsArray.forEach(([key, value]) => (sortedColumnsAsObject[key] = value));
    appSchema.definitions[lineItemId].properties = sortedColumnsAsObject;
}
/**
 * Determines the appropriate action definition name based on the provided annotation data and record type.
 *
 * @param {AnnotationSchemaContext} data - The schema data for the annotation, including its identifier and metadata.
 * @param {DataFieldAbstractTypes} annotationRecord - The record representing a data field or action within the annotation collection.
 * @param {boolean} isInnerAction - Whether the action belongs to the action group.
 * @returns {DefinitionName} The determined action definition name based on the provided data and group type.
 */
function determineActionDefinitionName(data, annotationRecord, isInnerAction) {
    const isGrouped = annotationRecord.$Type === "com.sap.vocabularies.UI.v1.DataFieldForActionGroup" /* UIAnnotationTypes.DataFieldForActionGroup */;
    let actionDefinitionName;
    if (data.annotation['term'] === "com.sap.vocabularies.UI.v1.Chart" /* UIAnnotationTerms.Chart */) {
        actionDefinitionName = isGrouped
            ? ux_specification_types_1.DefinitionName.ViewToolBarActionGroup
            : isInnerAction
                ? ux_specification_types_1.DefinitionName.ViewToolBarActionBase
                : ux_specification_types_1.DefinitionName.ChartToolBarAction;
    }
    else if (data.annotationId.startsWith(ux_specification_types_1.DefinitionName.LineItemsOfView)) {
        actionDefinitionName = isGrouped
            ? ux_specification_types_1.DefinitionName.ViewToolBarActionGroup
            : isInnerAction
                ? ux_specification_types_1.DefinitionName.ViewToolBarActionBase
                : ux_specification_types_1.DefinitionName.ViewToolBarAction;
    }
    else if (data.annotationId.startsWith(ux_specification_types_1.DefinitionName.LineItems)) {
        actionDefinitionName = isGrouped
            ? ux_specification_types_1.DefinitionName.ToolBarActionGroup
            : isInnerAction
                ? ux_specification_types_1.DefinitionName.ToolBarActionBase
                : ux_specification_types_1.DefinitionName.ToolBarAction;
    }
    else {
        actionDefinitionName = isGrouped
            ? ux_specification_types_1.DefinitionName.ObjectPageToolBarActionGroup
            : isInnerAction
                ? ux_specification_types_1.DefinitionName.ObjectPageToolBarActionBase
                : ux_specification_types_1.DefinitionName.ObjectPageToolBarAction;
    }
    return actionDefinitionName;
}
/**
 * Adds the schema tags 'keys' and 'target' to the schema definition of an action.
 *
 * @param {SchemaDefinition} actionDefinition - current action definition in the app schema
 * @param {DataFieldForActionAbstractTypes} lineItemRecord - actual record of the line item collection
 * @param {string} actionId - ID of the action, according to the lineItem definition
 */
function addTargetAndKeysOfAction(actionDefinition, lineItemRecord, actionId) {
    actionDefinition[ux_specification_types_1.SchemaTag.keys] = [];
    if (lineItemRecord[ux_specification_types_1.SchemaKeyName.semanticObject]) {
        actionDefinition[ux_specification_types_1.SchemaTag.keys].push({
            name: ux_specification_types_1.SchemaKeyName.semanticObject,
            value: lineItemRecord[ux_specification_types_1.SchemaKeyName.semanticObject]
        });
    }
    const { key, ns } = (0, common_1.getActionKey)(actionId, true);
    actionDefinition[ux_specification_types_1.SchemaTag.keys].push({ name: ux_specification_types_1.SchemaKeyName.action, value: key });
    if (ns) {
        actionDefinition[ux_specification_types_1.SchemaTag.target] = ns;
    }
}
/**
 * Adds an action record to a schema definition based on the provided annotation record definition parameters.
 * This involves defining action-specific properties, references, and metadata, and updating the schema accordingly.
 *
 * @param {AddAnnotationRecordDefinitionParams} annotationRecordParams - The parameters describing the annotation record definition, including data, actions, path, record, and index.
 * @param {boolean} [createRef] - A flag indicating whether to create a reference for the action definition.
 * @param {boolean} isInnerAction - Whether the action belongs to the action group.
 */
function addActionRecordToSchema(annotationRecordParams, createRef = true, isInnerAction = false) {
    const { schemaContext: data, actions: actionsDefinition, annotationPath, annotationRecord, index: recordIndex } = annotationRecordParams;
    // Action but not inline action
    if (annotationRecord.Determining === true) {
        return;
    }
    //no footer bar -> toolbar action
    const actionDefinitionName = determineActionDefinitionName(data, annotationRecord, isInnerAction);
    const description = (0, common_1.getDataFieldDescription)(annotationRecord, data.entityType);
    // Generate stable ID using strategy pattern to ensure correct key format for all action types.
    // This is crucial for IBN actions, which need the full key including SemanticObject and RequiresContext
    // (e.g., "DataFieldForIntentBasedNavigation::v4Freestyle::Inbound::RequiresContext")
    // to be properly matched when referenced in manifest-based action menus.
    const strategy = DataFieldStrategy_1.dataFieldStrategyContext.getStrategy(annotationRecord.$Type);
    const stableId = strategy.generateStableId(annotationRecord);
    // Extract Action property for keys generation (used by addTargetAndKeysOfAction).
    // For IBN actions, this is just the action name (e.g., "Inbound").
    // For bound actions, this may include namespace (e.g., "com.sap.Action/MyAction").
    const actionId = typeof annotationRecord.Action === 'string' ? annotationRecord.Action : annotationRecord.Action['path'];
    // Remove type prefix from stable ID for use in $ref generation.
    // This ensures getUniqueFacetDefinitionLink can properly detect duplicates based on full action identity,
    // not just the Action property. For example, two IBN actions with different SemanticObjects but same Action
    // should be treated as different actions, not duplicates.
    const stableIdWithoutPrefix = stableId.split('::').slice(1).join('::');
    // Create action definition with full stable ID as property key.
    // This allows manifest-based menus to correctly reference actions by their complete identity.
    const actionDefinition = (actionsDefinition.properties[stableId] = {
        description
    });
    if (createRef) {
        actionDefinition.$ref = (0, common_1.getUniqueFacetDefinitionLink)(actionDefinitionName, (0, common_1.prepareRef)(stableIdWithoutPrefix), data.appSchema, annotationPath);
    }
    else {
        actionDefinition.properties = {};
        actionDefinition.type = 'object';
        actionDefinition.isViewNode = true;
    }
    actionDefinition[ux_specification_types_1.SchemaTag.propertyIndex] = recordIndex;
    actionDefinition[ux_specification_types_1.SchemaTag.dataType] = (0, common_1.determineDataType)(annotationRecord);
    //keys
    addTargetAndKeysOfAction(actionDefinition, annotationRecord, actionId);
    actionDefinition[ux_specification_types_1.SchemaTag.annotationType] = annotationRecord.$Type;
    if (createRef) {
        // Using previously computed ref instead of generating new one to support action duplication
        const previouslyComputedRef = actionDefinition.$ref.replace(utils_1.DEFINITION_LINK_PREFIX, '');
        const action = (0, common_1.parseSchemaDefinition)(actionDefinitionName, previouslyComputedRef, data.appSchema, false);
        action[ux_specification_types_1.SchemaTag.annotationPath] = `${annotationPath}/${recordIndex}`;
    }
    else {
        actionDefinition[ux_specification_types_1.SchemaTag.annotationPath] = `${annotationPath}/${recordIndex}`;
    }
}
/**
 * Adds a group action record to the schema based on the given parameters. This involves determining
 * action definitions, creating or referencing schema properties, and setting metadata fields.
 *
 * @param {AddAnnotationRecordDefinitionParams} annotationRecordParams - Parameters that define the annotation record, including data, record, path, index, and actions.
 * @param {boolean} [createRef] - Indicates whether to create a reference for the action definition or to use an inline object.
 */
function addGroupActionRecordToSchema(annotationRecordParams, createRef = true) {
    const { schemaContext, annotationRecord, annotationPath, index, actions } = annotationRecordParams;
    //no footer bar -> toolbar action
    const actionDefinitionName = determineActionDefinitionName(schemaContext, annotationRecord);
    const propertyKey = (0, StableIdHelper_1.getStableIdPartFromDataField)(annotationRecord);
    let actionDefinition = {};
    if (createRef) {
        actions.properties[`${propertyKey}`] = {
            $ref: (0, common_1.getFacetDefinitionLink)(actionDefinitionName, (0, common_1.prepareRef)(propertyKey))
        };
        actionDefinition = (0, common_1.parseSchemaDefinition)(actionDefinitionName, (0, common_1.prepareRef)(propertyKey), schemaContext.appSchema);
    }
    else {
        actions.properties[`${propertyKey}`] = actionDefinition;
    }
    if (!actionDefinition.properties) {
        actionDefinition.properties = {};
    }
    actionDefinition.description = (0, common_1.getDataFieldDescription)(annotationRecord, schemaContext.entityType);
    actionDefinition.type = 'object';
    actionDefinition.isViewNode = true;
    actionDefinition[ux_specification_types_1.SchemaTag.propertyIndex] = index;
    actionDefinition[ux_specification_types_1.SchemaTag.dataType] = (0, common_1.determineDataType)(annotationRecord);
    actionDefinition[ux_specification_types_1.SchemaTag.annotationPath] = `${annotationPath}/${index}`;
    (annotationRecord['Actions'] ?? []).forEach((action, index) => {
        const innerActionParams = {
            ...annotationRecordParams,
            annotationRecord: action,
            annotationPath: `${actionDefinition[ux_specification_types_1.SchemaTag.annotationPath]}/Actions`,
            actions: actionDefinition,
            index
        };
        addActionRecordToSchema(innerActionParams, createRef, true);
    });
}
/**
 * Adds a line item action definition to the schema based on the provided parameters.
 * Handles determining which action to take based on specific conditions of the annotation record.
 *
 * @param {AddAnnotationRecordDefinitionParams} annotationRecordParams - An object containing the details of the annotation record and the inline definition.
 * @param {object} annotationRecordParams.annotationRecord - The annotation record to be processed, containing relevant properties.
 * @param {boolean} annotationRecordParams.annotationRecord.Inline - Indicates whether the record is inline.
 * @param {boolean} annotationRecordParams.annotationRecord.Determining - Indicates if the record is determining.
 * @param {boolean} annotationRecordParams.inlineDefinition - Indicates whether the inline definition is present.
 */
function addLineItemActionDefinition(annotationRecordParams) {
    const { annotationRecord, inlineDefinition } = annotationRecordParams;
    if (annotationRecord.Inline !== true || annotationRecord.Determining === true) {
        addActionRecordToSchema(annotationRecordParams, !inlineDefinition);
    }
    else {
        addLineItemRecordToSchema(annotationRecordParams);
    }
}
/**
 * Adds a line item record to the provided schema based on the definition parameters.
 *
 * @param {AddAnnotationRecordDefinitionParams} annotationRecordParams - The parameters defining the annotation data, schema information, and record.
 */
function addLineItemRecordToSchema(annotationRecordParams) {
    const { schemaContext, annotationRecord, inlineDefinition, annotationPath, index, definitionName } = annotationRecordParams;
    let columnDefinition;
    if (definitionName) {
        columnDefinition =
            annotationRecord.$Type === "com.sap.vocabularies.UI.v1.DataFieldForAction" /* UIAnnotationTypes.DataFieldForAction */
                ? `${definitionName}Action`
                : definitionName;
    }
    const lineItemId = schemaContext.annotationId;
    if (annotationRecord['Target']?.['value']) {
        const regex = `@${(0, utils_1.findAlias)(ux_specification_types_1.UIVOCABULARY, schemaContext.oDataServiceAVT)}.`;
        annotationRecord['Target']['value'] = annotationRecord['Target']['value'].replace(regex, '@');
    }
    const schemaKey = (0, StableIdHelper_1.getStableIdPartFromDataField)(annotationRecord).replace(/\//g, exports.FIELD_SEPARATOR);
    if (schemaKey) {
        let columnsDefinition = inlineDefinition;
        if (!columnsDefinition) {
            columnsDefinition = schemaContext.appSchema.definitions[lineItemId];
        }
        const properties = columnsDefinition.properties;
        const property = {
            description: (0, common_1.getDataFieldDescription)(annotationRecord, schemaContext.entityType)
        };
        if (inlineDefinition) {
            property.properties = {};
            property.type = 'object';
            property.isViewNode = true;
        }
        else if (columnDefinition) {
            property.$ref = utils_1.DEFINITION_LINK_PREFIX + columnDefinition;
        }
        property[ux_specification_types_1.SchemaTag.annotationType] = annotationRecord?.$Type;
        property[ux_specification_types_1.SchemaTag.annotationPath] = `${annotationPath}/${index}`;
        property[ux_specification_types_1.SchemaTag.propertyIndex] = index;
        property[ux_specification_types_1.SchemaTag.dataType] = (0, common_1.determineDataType)(annotationRecord);
        properties[schemaKey] = property;
        (0, common_1.addKeyToDefinition)(annotationRecord, properties, schemaKey);
    }
}
/**
 * Appends annotation record columns and actions to the app schema (for the list report, custom page or an object page section comprising a table).
 *
 * @param {AnnotationSchemaContext} data structure comprising
 * - appSchema - the app specific schema that shall get enhanced
 * - annotation - the UI.LineItem annotation, comprising all records
 * - entityType - the entity type as part of the AVT ConvertedMetadata
 * - oDataServiceAVT - complete service information, as returned by Annotation Vocabularies Tool
 * - annotationId - in case of OP the parameter must be passed to distinguish the OP tables;
 *                  in case of LR or ALP 'LineItems' is passed.
 * @param {Definition} actions - actions definition in schema, parent object
 * @param {string} lineItemPath - annotation path to the line item
 * @param {Definition | undefined} lineItemDefinition - line item definition
 * @param {string} columnDefinitionName - line item definition name
 */
function appendLineItemsToSchema(data, actions, lineItemPath, lineItemDefinition, columnDefinitionName = ux_specification_types_1.DefinitionName.TableColumn) {
    if (!data.annotation) {
        return;
    }
    let i = 0;
    data.annotation.forEach((lineItemRecord) => {
        const params = {
            schemaContext: data,
            actions,
            annotationPath: lineItemPath,
            annotationRecord: lineItemRecord,
            index: i,
            inlineDefinition: lineItemDefinition,
            definitionName: columnDefinitionName
        };
        const strategy = DataFieldStrategy_1.dataFieldStrategyContext.getStrategy(lineItemRecord.$Type);
        strategy.addLineItemDefinition(params);
        i++;
    });
}
/**
 * Adds the line item definition, columns and actions to the app schema (for the list report, custom page or an object page section comprising a table.
 *
 * @param data structure comprising
 * - isObjectPage - indicates that the page is an object page
 * - appSchema - the app specific schema that shall get enhanced
 * - annotation - the UI.LineItem annotation, comprising all records
 * - entityType - the entity type as part of the AVT ConvertedMetadata
 * - oDataServiceAVT - complete service information, as returned by Annotation Vocabularies Tool
 * - annotationId - in case of OP the parameter must be passed to distinguish the OP tables;
 *                in case of LR or ALP 'LineItems' is passed.
 * @param columnDefinitionName - name of the column definition, i.e. TableColumn or ObjectPageTableColumn
 * @param customColumnDefinitionName - name of the custom column definition
 * @param customActionDefinitionName - name of the custom action definition
 */
function addLineItemsTypeToSchema(data, columnDefinitionName = ux_specification_types_1.DefinitionName.TableColumn, customColumnDefinitionName = ux_specification_types_1.DefinitionName.CustomColumn, customActionDefinitionName = ux_specification_types_1.DefinitionName.CustomTableAction) {
    const { actions, lineItemPath } = (0, common_1.addCommonLineItemDefinitions)(data.isObjectPage, data.appSchema, data.annotation, data.entityType, data.annotationId, customColumnDefinitionName, customActionDefinitionName);
    appendLineItemsToSchema(data, actions, lineItemPath, undefined, columnDefinitionName);
}
/**
 * Adds an enum or value help to position > anchor of custom column.
 *
 * @param appSchema - app specific JSON schema
 * @param lineItemId - ID of the current line item in schema
 * @param v4Page - current page in manifest
 * @param positionName - ID of the definition of custom column position in schema
 * @param sectionIdInManifest - identifier of the current object page section in manifest
 */
function addFragmentEnumForAnchor(appSchema, lineItemId, v4Page, positionName = ux_specification_types_1.DefinitionName.Position, sectionIdInManifest) {
    const columns = appSchema.definitions[lineItemId]?.properties;
    const enumEntries = [];
    if (columns) {
        for (const key in columns) {
            if (key !== 'custom') {
                enumEntries.push(key);
            }
        }
    }
    const id = lineItemId !== ux_specification_types_1.DefinitionName.LineItems ? sectionIdInManifest : `@${"com.sap.vocabularies.UI.v1.LineItem" /* UIAnnotationTerms.LineItem */}`;
    const columnsInManifest = v4Page.options?.settings?.controlConfiguration?.[id]?.['columns'];
    if (columnsInManifest) {
        for (const key in columnsInManifest) {
            if (enumEntries.indexOf(key) === -1) {
                enumEntries.push(key);
            }
        }
    }
    if (enumEntries.length > 0) {
        const anchor = appSchema.definitions[positionName].properties.anchor;
        (0, common_1.addEnumToSchema)(enumEntries, anchor);
    }
}
/**
 * Adds an enum or value help to position > anchor of custom column or custom action.
 *
 * @param appSchema - app specific JSON schema
 * @param definitionId - definition key of actions
 * @param positionName - target definition to update
 */
function addEnumForActionAnchor(appSchema, definitionId, positionName = ux_specification_types_1.DefinitionName.CustomActionPosition) {
    (0, common_1.addDescriptiveEnumDefinition)(appSchema, appSchema.definitions[definitionId], {
        definition: positionName,
        property: 'anchor'
    }, {
        resolveDefinition: true
    });
}
/**
 * Adds an enum options of the columns to the table enableMassEdit properties.
 *
 * @param {Definition} appSchema - The schema of the application where the enums will be added.
 * @param {string} tableDefinitionName - The name of the table definition in the schema.
 * @param {string} columnsDefinitionName - The name of the columns definition used to extract column enumerations.
 * @param {EntityType} [entityType] - Optional entity type for additional context when retrieving column definitions.
 */
function addEnumForEnableMassEdit(appSchema, tableDefinitionName, columnsDefinitionName, entityType) {
    if (!entityType) {
        return;
    }
    const FIELD_PROPERTIES = ['visibleFields', 'ignoredFields'];
    const columnsDefinitions = appSchema.definitions?.[columnsDefinitionName];
    // Ignore check for hidden properties
    const filterOptions = {
        hidden: null
    };
    const fieldGroupsList = (0, propertiesForEdit_1.buildFieldGroupsList)(columnsDefinitions, entityType, filterOptions, 'single');
    const singleFieldList = (0, propertiesForEdit_1.buildSingleFieldList)(columnsDefinitions, entityType, filterOptions);
    const enableMassEditDefinition = (appSchema.definitions?.[tableDefinitionName]).properties?.['enableMassEdit'];
    const enableMassEditFields = enableMassEditDefinition.anyOf?.find((definition) => typeof definition === 'object' && definition && definition.type === 'object');
    for (const propertyKey in enableMassEditFields.properties) {
        if (FIELD_PROPERTIES.includes(propertyKey)) {
            const property = enableMassEditFields.properties[propertyKey];
            property.items = {
                type: 'string',
                // Indicate that items are tree structure
                uiType: 'tree',
                // Selection restricted within single group
                restrictGroups: true,
                // Only value from single group is allowed
                exclusiveGroups: ['FieldGroup'],
                // Allow annotation FieldGroup creation
                creation: 'FieldGroup',
                oneOf: [...fieldGroupsList, ...singleFieldList]
            };
            property.uniqueItems = true;
        }
    }
}
/**
 * Filters and retrieves column definitions that are not related to navigation properties
 * based on the provided application schema, line item ID, and entity type.
 *
 * @param appSchema The application schema definition containing the column definitions.
 * @param lineItemId The identifier of the line item whose column definitions are to be filtered.
 * @param entityType The entity type containing the navigation properties to filter against.
 * @returns A definition object containing column definitions that are not navigation properties.
 */
function getNonNavigationColumnDefinitions(appSchema, lineItemId, entityType) {
    const columns = appSchema.definitions[lineItemId]?.properties;
    const navigationProperties = new Set((entityType?.navigationProperties ?? []).map((np) => np.name));
    const filteredColumns = {};
    for (const key of Object.keys(columns)) {
        const column = columns[key];
        const propertyKeyValue = (column[ux_specification_types_1.SchemaTag.keys] ?? []).reduce((acc, currentKey) => {
            const { name, value } = currentKey;
            return name === ux_specification_types_1.SchemaKeyName.value || name === ux_specification_types_1.SchemaKeyName.target ? value : acc;
        }, '');
        const isNavigationProperty = propertyKeyValue.split('/').some((value) => navigationProperties.has(value));
        if (!isNavigationProperty) {
            filteredColumns[key] = columns[key];
        }
    }
    return filteredColumns;
}
/**
 * Initializes and modifies a specific schema definition for the creation mode of a line item.
 *
 * @param {Definition} appSchema - The base application schema containing definitions.
 * @param {string} lineItemId - The identifier for the line item in the schema.
 * @param {EntityType} entityType - The metadata entity type related to the line item.
 * @param {string} definitionName - The name of the general schema definition to be adapted.
 * @returns {string} The name of the newly created specific schema definition for the creation mode.
 */
function initializeCreationModeSchema(appSchema, lineItemId, entityType, definitionName) {
    const definitions = appSchema.definitions;
    const columnDefinitions = getNonNavigationColumnDefinitions(appSchema, lineItemId, entityType);
    const fieldGroupsList = entityType
        ? (0, propertiesForEdit_1.buildFieldGroupsList)(columnDefinitions, entityType, undefined, 'single', false)
        : [];
    const singleFieldList = entityType ? (0, propertiesForEdit_1.buildSingleFieldList)(columnDefinitions, entityType) : [];
    const specificCreationModeDefinitionName = (0, common_1.getFacetDefinitionKey)(definitionName, lineItemId);
    const specificCreationModeDefinition = (definitions[specificCreationModeDefinitionName] = structuredClone(definitions[definitionName]));
    const creationFieldsDefinition = specificCreationModeDefinition.properties?.[ux_specification_types_1.PropertyName.creationFields];
    const creationFieldsDefinitionItems = creationFieldsDefinition.items;
    creationFieldsDefinitionItems['oneOf'] = [...fieldGroupsList, ...singleFieldList];
    // Indicate that items are tree structure
    creationFieldsDefinitionItems['uiType'] = 'tree';
    // Allow annotation FieldGroup creation
    creationFieldsDefinitionItems['creation'] = 'FieldGroup';
    // Selection restricted within single group
    creationFieldsDefinitionItems['restrictGroups'] = true;
    // Only value from single group is allowed
    creationFieldsDefinitionItems['exclusiveGroups'] = ['FieldGroup'];
    // Entries of one of are translatable using 'service' i18n bundle
    creationFieldsDefinitionItems.i18nBundle = 'service';
    creationFieldsDefinition.uniqueItems = true;
    return specificCreationModeDefinitionName;
}
/**
 * Updates the schema for specific creation fields in the provided application schema.
 * This method dynamically modifies schemas of particular table types to include specific definitions for creation fields.
 *
 * @param appSchema The application schema to update, represented as a Definition object.
 * @param entityType The entity type used for determining field groups.
 * @param definitionName The name of the schema definition that serves as the base for the specific creation mode schema.
 * @param lineItemId The identifier of the line item related to the creation fields to be updated in the schema.
 */
function addOneOfForCreationFields(appSchema, entityType, definitionName, lineItemId) {
    const creationModeDefinitionName = initializeCreationModeSchema(appSchema, lineItemId, entityType, definitionName);
    // Replace the generic creationMode definition reference with the facet-specific one
    const tableDefinition = appSchema.definitions[`${ux_specification_types_1.DefinitionName.ObjectPageTable}<${lineItemId}>`];
    const creationModeProperty = tableDefinition?.properties?.creationMode;
    if (creationModeProperty?.$ref) {
        creationModeProperty.$ref = creationModeProperty.$ref.replace(definitionName, creationModeDefinitionName);
    }
}
/**
 * Initializes the personalization schema for a specific table by creating section-specific definitions.
 * This handles Object Page tables with multiple sections where personalization settings need to be section-specific.
 *
 * @param appSchema The application schema to update, represented as a Definition object.
 * @param definitionName The name of the schema definition that serves as the base for the specific personalization schema (e.g., DefinitionName.PersonalizationTypeOP).
 * @param idOrKey The lineItemId or viewKey used to construct the facet definition keys.
 * @param tableDefinitionName The base table definition name (e.g., DefinitionName.ObjectPageTable or DefinitionName.LRTableView).
 * @returns The name of the newly created section-specific personalization definition.
 */
function initializePersonalizationSchema(appSchema, definitionName, idOrKey, tableDefinitionName) {
    // Calculate facet definition keys inside the function
    const specificPersonalizationDefinitionName = (0, common_1.getFacetDefinitionKey)(definitionName, idOrKey);
    const tableDefinitionKey = (0, common_1.getFacetDefinitionKey)(tableDefinitionName, idOrKey);
    // Clone the base personalization definition for this specific section
    if (appSchema.definitions) {
        appSchema.definitions[specificPersonalizationDefinitionName] = structuredClone(appSchema.definitions[definitionName]);
    }
    // Replace the generic personalization definition reference in the table with the facet-specific one
    const tableDefinition = appSchema.definitions?.[tableDefinitionKey];
    const personalizationProperty = tableDefinition?.properties?.personalization;
    // The personalization property uses anyOf, so we need to update the $ref within it
    if (personalizationProperty?.anyOf) {
        personalizationProperty.anyOf.forEach((item) => {
            if (typeof item === 'object' && item.$ref) {
                item.$ref = item.$ref.replace(definitionName, specificPersonalizationDefinitionName);
            }
        });
    }
    return specificPersonalizationDefinitionName;
}
/**
 * Find the relevant V4 page under the routing targets of manifest,json.
 *
 * @param pages - list of all pages in manifest
 * @param templateName - search criterion: template name
 * @param entitySet - search criterion: entity set object from AVT
 * @param contextPath - search criterion: contextPath
 * @returns the page definition in manifest (if found)
 */
function findPageV4(pages, templateName, entitySet, contextPath) {
    let v4Page;
    for (const i in pages) {
        const pageSettings = pages[i].options?.settings || {};
        const isConnectionMatches = (contextPath && pageSettings.contextPath === contextPath) ||
            (entitySet?.name && pageSettings.entitySet === entitySet.name);
        if (isConnectionMatches &&
            ((0, common_1.compareTemplateNames)(templateName, pages[i].name) ||
                (templateName === ux_specification_types_1.v4.FE_TEMPLATE_V4_ALP && pages[i].name === ux_specification_types_1.v4.FE_TEMPLATE_V4_LIST_REPORT))) {
            v4Page = pages[i];
            break;
        }
    }
    return v4Page;
}
/**
 * Depending on the template type, strip down the app schema so that only the relevant views are part of it.
 *
 * @param {Definition} appSchema - app specific JSON schema, to be adjusted
 * @param templateType - template type of the current page
 * @returns the right definition name for the table in the schema
 */
function alignSchemaWithTemplateType(appSchema, templateType) {
    let tableDefinitionName;
    const tableDefinition = appSchema.properties[ux_specification_types_1.PropertyName.table];
    if (templateType === ux_specification_types_1.TemplateType.AnalyticalListPageV4) {
        tableDefinitionName = ux_specification_types_1.DefinitionName.ALPTableView;
        delete tableDefinition.anyOf;
        tableDefinition.$ref = `${utils_1.DEFINITION_LINK_PREFIX}${ux_specification_types_1.DefinitionName.ALPTableView}`;
        delete appSchema.definitions[ux_specification_types_1.DefinitionName.Table];
        delete appSchema.definitions[ux_specification_types_1.DefinitionName.LRChartView];
        delete appSchema.definitions[ux_specification_types_1.DefinitionName.LRTableView];
        delete appSchema.definitions[ux_specification_types_1.DefinitionName.MultiTableModeV4];
    }
    else if (templateType === ux_specification_types_1.TemplateType.ListReportObjectPageV4) {
        tableDefinitionName = ux_specification_types_1.DefinitionName.Table;
        delete tableDefinition.anyOf;
        tableDefinition.$ref = `${utils_1.DEFINITION_LINK_PREFIX}${ux_specification_types_1.DefinitionName.Table}`;
        appSchema.definitions[tableDefinitionName].properties[ux_specification_types_1.PropertyName.annotationPath][ux_specification_types_1.SchemaTag.hidden] = true;
        delete appSchema.definitions[ux_specification_types_1.DefinitionName.ALPTableView];
        appSchema.properties[ux_specification_types_1.PropertyName.defaultPath][ux_specification_types_1.SchemaTag.hidden] = true;
    }
    return tableDefinitionName;
}
/**
 * Common function for enhancing LineItems, FieldGroups and Custom Section definitions of app schema by custom action definitions.
 *
 * @param appSchema The application schema where the custom action definition will be added.
 * @param v4Page The V4 page configuration object that contains metadata and settings for the page.
 * @param logger A logging instance for logging warnings or errors during the processing.
 * @param customActionDefinitionName The name of the custom action definition to be added. Defaults to `DefinitionName.CustomTableAction`.
 * @param sectionDefinitionName The name of the section definition within the schema to which the action definition will be added.
 * @param sectionIdInManifest Optional: The section ID in the manifest, used to derive the configuration ID.
 * @param lineItemId Optional: The ID of the line item, if applicable.
 */
function addCustomActionDefinition(appSchema, v4Page, logger, customActionDefinitionName = ux_specification_types_1.DefinitionName.CustomTableAction, sectionDefinitionName = '', sectionIdInManifest = '', lineItemId) {
    const definition = appSchema.definitions[sectionDefinitionName];
    if (!definition) {
        return;
    }
    // Sort existing actions
    let sortedActionsArray = sortPropertiesRecords(definition);
    let configId;
    if (sectionIdInManifest) {
        if (customActionDefinitionName.indexOf(ux_specification_types_1.DefinitionName.CustomFormActionOP) !== -1) {
            configId = sectionIdInManifest.replace(/::/g, '#');
        }
        else {
            configId = sectionIdInManifest.replace(/::/g, '/');
        }
    }
    else {
        if (customActionDefinitionName === ux_specification_types_1.DefinitionName.CustomFormActionOP) {
            configId = `@${"com.sap.vocabularies.UI.v1.FieldGroup" /* UIAnnotationTerms.FieldGroup */}`;
        }
        else {
            configId = `@${"com.sap.vocabularies.UI.v1.LineItem" /* UIAnnotationTerms.LineItem */}`;
        }
    }
    let actions;
    if (sectionDefinitionName.includes(ux_specification_types_1.DefinitionName.ObjectPageCustomSectionActions)) {
        actions = v4Page.options?.settings?.content?.body?.sections?.[configId]?.['actions'];
    }
    else {
        actions = v4Page.options?.settings?.controlConfiguration?.[configId]?.['actions'];
    }
    if (actions) {
        // Add custom actions and custom action menus to the actions array
        const parameters = {
            extensions: actions,
            appSchema,
            customActionDefinitionName,
            initialActionDetailsList: sortedActionsArray,
            logger,
            lineItemId
        };
        sortedActionsArray = addActionsToArray(parameters);
    }
    // Adjust propertyIndex
    if (sortedActionsArray) {
        for (let index = 0; index < sortedActionsArray.length; index++) {
            sortedActionsArray[index][1][ux_specification_types_1.SchemaTag.propertyIndex] = index;
        }
        const sortedActionsAsObject = {};
        sortedActionsArray.forEach(([key, value]) => (sortedActionsAsObject[key] = value));
        appSchema.definitions[sectionDefinitionName].properties = sortedActionsAsObject;
    }
}
/**
 * Common function for enhancing header and footer definitions of app schema by custom action definitions.
 *
 * @param appSchema - app-specific JSON schema
 * @param v4Page - actual page in the manifest
 * @param logger - logger for error messages
 * @param customActionDefinitionName - definition name of custom action
 * @param sectionId - identifier of the current object page section in schema
 */
function addHeaderFooterCustomActionDefinition(appSchema, v4Page, logger, customActionDefinitionName = ux_specification_types_1.DefinitionName.CustomHeaderActionOP, sectionId = '') {
    const definition = appSchema.definitions[sectionId];
    if (!definition) {
        return;
    }
    // Sort existing actions
    let sortedActionsArray = sortPropertiesRecords(definition);
    let actions;
    if ([ux_specification_types_1.DefinitionName.CustomHeaderAction, ux_specification_types_1.DefinitionName.CustomHeaderActionOP].includes(customActionDefinitionName)) {
        actions = v4Page.options?.settings?.content?.header?.['actions'];
    }
    else {
        actions = v4Page.options?.settings?.content?.footer?.['actions'];
    }
    if (actions) {
        // Add custom actions and custom action menus to the actions array
        const parameters = {
            extensions: actions,
            appSchema,
            customActionDefinitionName,
            initialActionDetailsList: sortedActionsArray,
            logger
        };
        sortedActionsArray = addActionsToArray(parameters);
    }
    // Adjust propertyIndex
    const sortedActionsAsObject = {};
    if (sortedActionsArray) {
        for (let index = 0; index < sortedActionsArray.length; index++) {
            sortedActionsArray[index][1][ux_specification_types_1.SchemaTag.propertyIndex] = index;
        }
        sortedActionsArray.forEach(([key, value]) => (sortedActionsAsObject[key] = value));
    }
    appSchema.definitions[sectionId].properties = sortedActionsAsObject;
}
/**
 * Adds actions and custom action menus to an array of initial actions using specified parameters.
 *
 * @param params - An object containing the parameters for adding actions to the array:
 *   - actions: The original actions to be processed.
 *   - initialActionsArray: The array to which actions should be added.
 *   - customActionDefinitionName: The name of the custom action definition.
 *   - appSchema: The application schema used for processing.
 *   - logger: An object used for logging information during processing.
 * @returns An array of action details after processing and adding extensions.
 */
function addActionsToArray(params) {
    const { extensions, initialActionDetailsList, customActionDefinitionName, logger } = params;
    // Add custom actions
    let actionDetailsList = addExtensionToArray(extensions, initialActionDetailsList, customActionDefinitionName, [ux_specification_types_1.TemplatePropertyName.Press, ux_specification_types_1.TemplatePropertyName.Menu], logger, false, params.appSchema);
    const processActionMenusParams = {
        ...params,
        initialActionDetailsList: actionDetailsList
    };
    actionDetailsList = processActionMenus(processActionMenusParams);
    return actionDetailsList;
}
/**
 * Processes action menus by identifying menu-type actions within the provided extensions and updating the list of action details accordingly.
 *
 * @param {ActionProcessingParameters} params - An object containing required parameters for processing action menus. It includes:
 *   - extensions: A map of extension data where the keys are extension IDs and the values are extension definitions.
 *   - initialActionDetailsList: A list of action details, each represented as a tuple with the action ID and action definition.
 * @returns {Array} An updated list of action details after processing any menu-type actions.
 */
function processActionMenus(params) {
    const { extensions, initialActionDetailsList } = params;
    let actionDetailsList = [...initialActionDetailsList];
    // Currently, all actions were initialized as custom actions. We need to check if the action is a menu and if so,
    // we need to process it differently.
    for (const extensionId in extensions) {
        const extension = extensions[extensionId];
        const actionDetails = actionDetailsList.find(([actionId]) => actionId === extensionId);
        if (Object.keys(extension).includes(ux_specification_types_1.PropertyName.menu) && actionDetails) {
            const [, actionDefinition] = actionDetails;
            actionDetailsList = processActionMenu(params, extension, actionDefinition, actionDetailsList, extensionId);
        }
    }
    return actionDetailsList;
}
/**
 * Retrieves a reference to a menu entry definition based on the provided parameters.
 *
 * @param {string} $ref - The reference string used to identify the menu entry definition.
 * @param {string} actionMenuDefinitionName - The name of the action menu definition.
 * @param {Definition} appSchema - The application schema containing the definitions.
 * @returns {string} The constructed link or reference for the specified menu entry definition.
 */
function getMenuEntryDefinitionLink($ref, actionMenuDefinitionName, appSchema) {
    const definitionName = (0, utils_1.getDefinitionKey)($ref);
    const { baseDefinitionName, definitionQualifier } = parseDefinitionName(definitionName);
    // incorrect definition name - probably a reference to an intent based navigation action
    if (baseDefinitionName.includes(exports.FIELD_SEPARATOR)) {
        return $ref;
    }
    const menuEntryDefinitionQualifier = definitionQualifier
        ? `${actionMenuDefinitionName}${exports.FIELD_SEPARATOR}${definitionQualifier}`
        : actionMenuDefinitionName;
    const menuEntryDefinitionRef = (0, common_1.getFacetDefinitionKey)(baseDefinitionName, menuEntryDefinitionQualifier);
    const menuEntryDefinitionLink = (0, common_1.getFacetDefinitionLink)(baseDefinitionName, menuEntryDefinitionQualifier);
    const menuEntryDefinition = appSchema.definitions[menuEntryDefinitionRef];
    if (!menuEntryDefinition) {
        const originalDefinition = appSchema.definitions[definitionName];
        const menuEntryDefinition = (0, common_1.parseSchemaDefinition)(baseDefinitionName, menuEntryDefinitionQualifier, appSchema);
        for (const originalDefinitionKey of Object.keys(originalDefinition)) {
            if (menuEntryDefinition[originalDefinitionKey] === undefined) {
                menuEntryDefinition[originalDefinitionKey] = originalDefinition[originalDefinitionKey];
            }
        }
        // some properties doesn't make sense for menu entries - hide them instead of deleting
        // NOTE: This is a necessary trade-off. When moving IBN actions in and out of menus,
        // we keep these properties visible in the manifest but hidden in the UI.
        // Unfortunately, this is required - otherwise, when renaming actions, we end up with
        // orphaned entries in the manifest.
        if (menuEntryDefinition.properties[ux_specification_types_1.PropertyName.position]) {
            menuEntryDefinition.properties[ux_specification_types_1.PropertyName.position][ux_specification_types_1.SchemaTag.hidden] = true;
        }
        if (menuEntryDefinition.properties[ux_specification_types_1.PropertyName.overflowGroup]) {
            menuEntryDefinition.properties[ux_specification_types_1.PropertyName.overflowGroup][ux_specification_types_1.SchemaTag.hidden] = true;
        }
        if (menuEntryDefinition.properties[ux_specification_types_1.PropertyName.priority]) {
            menuEntryDefinition.properties[ux_specification_types_1.PropertyName.priority][ux_specification_types_1.SchemaTag.hidden] = true;
        }
    }
    return menuEntryDefinitionLink;
}
/**
 * Determines the appropriate action definition names based on a custom action definition name.
 * Returns menu definition names and the action name for menu entries.
 *
 * @param {string} customActionDefinitionName - The custom action definition name to be parsed and matched.
 * @returns {ActionMenuDefinitionName} The corresponding action definition names, or a default value if not found.
 */
function determineActionDefinitionNames(customActionDefinitionName) {
    const { baseDefinitionName } = parseDefinitionName(customActionDefinitionName);
    return actionMenuDefinitionNameMap.get(baseDefinitionName) ?? LR_ACTION_MENU_DEFINITION;
}
/**
 * Parses a given definition name into its base name and optional qualifier.
 *
 * @param definitionName The definition name to parse. This should be a string in the format "BaseName" or "BaseName<Qualifier>".
 * @returns An object containing the base definition name and, if applicable, its qualifier.
 * If there is no qualifier, it will return undefined for the qualifier field.
 */
function parseDefinitionName(definitionName = '') {
    const match = /^([^<]*)(?:<(.*)>)?$/.exec(definitionName);
    if (!match) {
        return { baseDefinitionName: definitionName };
    }
    const [, baseDefinitionName, definitionQualifier] = match;
    return {
        baseDefinitionName: baseDefinitionName,
        definitionQualifier: definitionQualifier ?? undefined
    };
}
/**
 * Updates the custom menu's default action to include a list of valid options based on the provided menu schema.
 *
 * @param {Definition} customMenuDefinition - The definition object for the custom menu, where the default action will be updated.
 * @param {Record<string, SchemaDefinition>} [menu] - A mapping of menu items to their schema definitions, used to generate valid options for the default action.
 */
function setCustomMenuDefaultAction(customMenuDefinition, menu = {}) {
    if (customMenuDefinition.properties[ux_specification_types_1.PropertyName.defaultAction]) {
        const menuOneOfValues = Object.keys(menu).map((value) => ({
            const: value,
            description: menu[value].description
        }));
        customMenuDefinition.properties[ux_specification_types_1.PropertyName.defaultAction][ux_specification_types_1.PropertyName.oneOf] = menuOneOfValues;
    }
}
/**
 * Retrieves the description of a menu entry from its schema definition.
 *
 * @param menuEntryDefinition The schema definition of the menu entry, including its reference information.
 * @param appSchema The complete application schema containing definitions for the menu entries.
 * @returns The description of the menu entry if found, or an empty string if no description is available.
 */
function getMenuEntryDescription(menuEntryDefinition, appSchema) {
    const definitionKey = (0, utils_1.getDefinitionKey)(menuEntryDefinition.$ref);
    const definition = appSchema.definitions[definitionKey];
    return menuEntryDefinition?.description ?? definition?.description ?? '';
}
/**
 * Processes an action menu for a custom extension, updating the schema definition and action details list accordingly.
 *
 * @param {ActionProcessingParameters} params The parameters required for action menu processing, including `actionMenu`, `actionMenuActions`, `appSchema`, and `lineItemId`.
 * @param {CustomExtension} extension The custom extension that contains the action menu configuration.
 * @param {SchemaDefinition} actionDefinition The schema definition object to be populated with details of the custom action menu.
 * @param {ActionDetails[]} initialDetailsList The initial list of action details to be processed and modified.
 * @param {string} extensionId The unique identifier for the extension being processed.
 * @returns {ActionDetails[] | undefined} Returns the updated list of action details if processing occurs; otherwise, returns `undefined` if the extension is not a custom action menu.
 */
function processActionMenu(params, extension, actionDefinition, initialDetailsList, extensionId) {
    // if the extension isn't a custom action menu, don't do anything
    if (!(ux_specification_types_1.TemplatePropertyName.Menu in extension)) {
        return undefined;
    }
    const { customActionDefinitionName, appSchema, lineItemId } = params;
    const detailsList = [...initialDetailsList];
    const { actionMenuDefinitionName, actionMenuActionsDefinitionName } = determineActionDefinitionNames(customActionDefinitionName);
    // Prepare submenu entries to be processed
    const menu = extension[ux_specification_types_1.TemplatePropertyName.Menu] ?? [];
    const properties = {};
    // For forms and tables, the id should contain both the section and action names
    const extendedExtensionId = lineItemId ? `${extensionId}${exports.FIELD_SEPARATOR}${lineItemId}` : extensionId;
    // Create the specific action menu definition in the schema based on the generic action menu definition
    const customMenuDefinition = (0, common_1.parseSchemaDefinition)(actionMenuDefinitionName, extendedExtensionId, appSchema);
    actionDefinition.$ref = (0, common_1.getFacetDefinitionLink)(actionMenuDefinitionName, extendedExtensionId);
    actionDefinition[ux_specification_types_1.SchemaTag.actionType] = ux_specification_types_1.ActionType.CustomMenu;
    const customMenuActionsDefinition = (0, common_1.parseSchemaDefinition)(actionMenuActionsDefinitionName, extendedExtensionId, appSchema);
    const customMenuActionsDefinitionLink = (0, common_1.getFacetDefinitionLink)(actionMenuActionsDefinitionName, extendedExtensionId);
    // for manifest based menu actions, we need to store the menu entries from the manifest
    actionDefinition[ux_specification_types_1.PropertyName.menu] = menu;
    // Add the menu items to the action menu definition
    menu.forEach((menuEntry, propertyIndex) => {
        const sortedExtensionIndex = detailsList.findIndex(([entry]) => entry === menuEntry);
        // If the list of actions contains the menu entry, add the menu entry to the properties of the custom action menu
        // and remove the extension reference from the list of actions
        if (sortedExtensionIndex !== -1) {
            const menuEntryDefinition = detailsList[sortedExtensionIndex][1];
            menuEntryDefinition.propertyIndex = propertyIndex;
            menuEntryDefinition.isViewNode = true;
            menuEntryDefinition.$ref = getMenuEntryDefinitionLink(menuEntryDefinition.$ref, actionMenuDefinitionName, appSchema);
            menuEntryDefinition.description = getMenuEntryDescription(menuEntryDefinition, appSchema);
            properties[menuEntry] = menuEntryDefinition;
            detailsList.splice(sortedExtensionIndex, 1);
        }
    });
    customMenuDefinition.description = actionDefinition.description;
    // Update the properties of the custom action menu definition with the menu items
    const actions = customMenuDefinition.properties[ux_specification_types_1.PropertyName.actions];
    setCustomMenuDefaultAction(customMenuDefinition, properties);
    customMenuActionsDefinition.properties = properties;
    actions.$ref = customMenuActionsDefinitionLink;
    return detailsList;
}
/**
 * Method returns page from manifest by passed page key.
 *
 * @param manifest Object from manifest.json.
 * @param pageId Page id.
 * @param logger Logger to report error of unexisting page.
 * @returns Manifest's target page.
 */
function getManifestPage(manifest, pageId, logger) {
    const pages = (0, utils_1.getJSONPropertyByPath)(manifest, [
        ux_specification_types_1.ManifestSection.ui5,
        'routing',
        'targets'
    ]);
    if (!pages) {
        (0, __1.log)(logger, {
            severity: "error" /* LogSeverity.Error */,
            message: i18next_1.default.t('NOTARGETS'),
            location: {
                path: ux_specification_types_1.MANIFESTPATH,
                range: [ux_specification_types_1.ManifestSection.ui5, 'routing']
            }
        });
        return;
    }
    const page = (0, utils_1.getJSONPropertyByPath)(pages, [pageId]);
    if (!page) {
        (0, __1.log)(logger, {
            severity: "error" /* LogSeverity.Error */,
            message: i18next_1.default.t('NOROUTINGID', { routingId: pageId }),
            location: {
                path: ux_specification_types_1.MANIFESTPATH,
                range: [ux_specification_types_1.ManifestSection.ui5, 'routing']
            }
        });
        return;
    }
    return page;
}
/**
 * Method finds custom view/section XML file for passed page.
 *
 * @param files All extension files.
 * @param page - the page object containing routing target options
 * @param manifest Object from manifest.json.
 * @param xmlType XML file type - view or fragment.
 * @param extensionName Extension namespace or path to extension name in page object.
 * @returns Custom page's data containing XML file.
 */
function getPageCustomExtensionFile(files, page, manifest, xmlType, extensionName) {
    if (extensionName) {
        const appId = manifest['sap.app']?.id || '';
        const fileExtension = 'xml';
        if (extensionName.startsWith(appId)) {
            // Remove app id from extension name
            extensionName = extensionName.replace(appId, '');
        }
        const fileParts = extensionName.split('.');
        let baseName = fileParts.pop();
        baseName = `${baseName}.${xmlType}.${fileExtension}`;
        const extPath = (0, path_1.join)(...fileParts, baseName);
        const file = (files ?? []).find((file) => file.dataSourceUri.endsWith(extPath));
        if (file) {
            const settings = page.options?.settings;
            return {
                file,
                relativeFilePath: extPath,
                connection: {
                    contextPath: settings?.contextPath,
                    entitySet: settings?.entitySet
                }
            };
        }
    }
    return undefined;
}
/**
 * Method finds view XML file for passed page.
 *
 * @param files All view files.
 * @param pageId Page id.
 * @param manifest Object from manifest.json.
 * @param logger Logger.
 * @returns Custom page's data containing view XML file.
 */
function getPageCustomViewFile(files, pageId, manifest, logger) {
    const page = getManifestPage(manifest, pageId, logger);
    const viewName = (0, utils_1.getJSONPropertyByPath)(page, ['options', 'settings', 'viewName']);
    return page && typeof viewName === 'string'
        ? getPageCustomExtensionFile(files, page, manifest, 'view', viewName)
        : undefined;
}
/**
 * Adds custom section or custom sub section definitions to app schema.
 *
 * @param appSchema - app-specific JSON schema
 * @param definition - sections definition
 * @param v4Page - actual page in the manifest
 * @param customSectionRef - value of custom section ref
 * @param isMergedSections - are merged section used
 * @param facetKey - facet key
 * @param logger - logger for error messages
 */
function addCustomSectionDefinition(appSchema, definition, v4Page, customSectionRef, isMergedSections, facetKey, logger) {
    let sortedSectionsArray = sortPropertiesRecords(definition);
    const annotationSectionsMap = new Map();
    sortedSectionsArray = sortedSectionsArray.map((annotationSection) => {
        const entry = (0, common_1.getEnumEntry)(appSchema, annotationSection[1], annotationSection[0], {
            keyProperties: [ux_specification_types_1.SchemaKeyName.id],
            descriptionAsFallback: true,
            resolveDefinition: true
        });
        annotationSectionsMap.set(entry.const, annotationSection[0]);
        return [entry.const, annotationSection[1]];
    });
    if (facetKey) {
        // check for already defined section
        const section = v4Page.options?.settings?.content?.body?.sections?.[facetKey];
        if (section && 'subSections' in section) {
            // add extension to subsection of section
            sortedSectionsArray = addExtensionToArray(section.subSections, sortedSectionsArray, customSectionRef, ux_specification_types_1.TemplatePropertyName.Template, logger, isMergedSections);
        }
    }
    else if (customSectionRef.includes(ux_specification_types_1.DefinitionName.ObjectPageCustomHeaderSectionFragment)) {
        // header custom section extensions
        const header = v4Page.options?.settings?.content?.header;
        if (header && 'facets' in header) {
            // add extension to header section
            sortedSectionsArray = addExtensionToArray(header.facets, sortedSectionsArray, customSectionRef, ux_specification_types_1.TemplatePropertyName.Template, logger, isMergedSections);
        }
    }
    else {
        // body custom section extensions — accept both fragment-based (template) and
        // embeddedComponent-based (e.g. HDM attachment reuse) sections
        const sections = v4Page.options?.settings?.content?.body?.sections;
        if (sections) {
            sortedSectionsArray = addExtensionToArray(sections, sortedSectionsArray, customSectionRef, [ux_specification_types_1.TemplatePropertyName.Template, ux_specification_types_1.TemplatePropertyName.EmbeddedComponent], logger, isMergedSections);
        }
    }
    // Adjust propertyIndex
    for (let index = 0; index < sortedSectionsArray.length; index++) {
        sortedSectionsArray[index][1][ux_specification_types_1.SchemaTag.propertyIndex] = index;
    }
    const sortedActionsAsObject = {};
    sortedSectionsArray.forEach(([key, value]) => {
        key = annotationSectionsMap.get(key) || key;
        sortedActionsAsObject[key] = value;
    });
    definition.properties = sortedActionsAsObject;
}
/**
 * Method adds custom filter fields from manifest to schema properties.
 *
 * @param appSchema - app-specific JSON schema
 * @param definition - filter fields definition
 * @param v4Page - actual page in the manifest
 */
function addCustomFilterFieldDefinition(appSchema, definition, v4Page) {
    let sortedFilterFieldsArray = sortPropertiesRecords(definition);
    const annotationSectionsMap = new Map();
    sortedFilterFieldsArray = sortedFilterFieldsArray.map((annotationSection) => {
        const entry = (0, common_1.getEnumEntry)(appSchema, annotationSection[1], annotationSection[0], {
            keyProperties: [ux_specification_types_1.SchemaKeyName.value],
            descriptionAsFallback: true,
            resolveDefinition: true
        });
        annotationSectionsMap.set(entry.const, annotationSection[0]);
        return [entry.const, annotationSection[1]];
    });
    // check for already defined section
    const selectionFieldsEntry = v4Page.options?.settings?.controlConfiguration?.[`@${"com.sap.vocabularies.UI.v1.SelectionFields" /* UIAnnotationTerms.SelectionFields */}`];
    if (typeof selectionFieldsEntry === 'object' && 'filterFields' in selectionFieldsEntry) {
        // add extension to subsection of section
        sortedFilterFieldsArray = addExtensionToArray(selectionFieldsEntry.filterFields, sortedFilterFieldsArray, ux_specification_types_1.DefinitionName.CustomFilterField, ux_specification_types_1.TemplatePropertyName.Template);
    }
    // Adjust propertyIndex
    for (let index = 0; index < sortedFilterFieldsArray.length; index++) {
        sortedFilterFieldsArray[index][1][ux_specification_types_1.SchemaTag.propertyIndex] = index;
    }
    const sortedFilterFieldsAsObject = {};
    sortedFilterFieldsArray.forEach(([key, value]) => {
        key = annotationSectionsMap.get(key) || key;
        sortedFilterFieldsAsObject[key] = value;
    });
    definition.properties = sortedFilterFieldsAsObject;
}
/**
 * Imports action settings into a configuration object and applies necessary transformations based on provided parameters.
 *
 * @param {object} importActionParams - Parameters required for importing action settings, including schema, factory, and action specifications.
 * @param {string} key - The key used to identify the specific configuration or schema definition.
 * @param {boolean} [keepEmptyAction] - Determines whether to keep empty actions in the configuration object.
 */
function importActionToConfigurationObject(importActionParams, key, keepEmptyAction = false) {
    const { appSchema, factory, actionsInConfig, actionKey, manifest, routingId, manifestSectionId, targetAnnotation } = importActionParams;
    const schemaPropertyName = (key ?? '').split('<')[0];
    const metaInstance = factory.createInstance(ux_specification_types_1.PageTypeV4.ObjectPage, schemaPropertyName);
    const breadcrumbs = [actionKey.replace('/', '??')];
    if (targetAnnotation) {
        breadcrumbs.push(targetAnnotation);
    }
    if (metaInstance) {
        const action = (actionsInConfig[actionKey] = metaInstance);
        (0, utils_2.importSettingsOfObject)(action, manifest, appSchema.definitions[key], routingId, breadcrumbs, manifestSectionId);
        if (!keepEmptyAction) {
            (0, common_1.removeEmptyStructure)(actionsInConfig, actionKey);
        }
    }
}
/**
 * Imports the settings of a single action reference into the configuration object.
 *
 * @param {ImportActionSettingsOfReferenceParams} importActionParams - The parameters required for importing action settings, including property definitions.
 * @param {Definition} definitionOfReference - The reference definition object used to resolve the action.
 */
function importSingleActionSettingsOfReference(importActionParams, definitionOfReference) {
    const { actionPropertyDefinition } = importActionParams;
    const key = (0, utils_1.getDefinitionKey)(definitionOfReference.$ref || actionPropertyDefinition.$ref);
    importActionToConfigurationObject(importActionParams, key);
}
/**
 * Imports and processes the settings for a group of actions based on a reference definition.
 *
 * @param {ImportActionSettingsOfReferenceParams} importActionParams - Parameters containing details required to import group action settings, such as action property definitions, application schema, action configurations, and action keys.
 * @param {Definition} definitionOfReference - The definition object that serves as the reference for importing group action settings.
 * @param {string} definitionKey - Optional, the reference definition key from which the action group settings are imported.
 */
function importGroupActionSettingsOfReference(importActionParams, definitionOfReference, definitionKey) {
    const { actionPropertyDefinition, appSchema, actionsInConfig, actionKey } = importActionParams;
    let resolvedDefinitionOfReference = definitionOfReference;
    let resolvedDefinitionKey = (0, utils_1.getDefinitionKey)(resolvedDefinitionOfReference.$ref ?? actionPropertyDefinition.$ref);
    /* Some action groups (e.g. form group actions) have wrapper definition in the schema, in such cases we need to use the referenced definition. */
    const hasActionGroupWrapper = (0, utils_1.getDefinitionKey)(definitionOfReference.$ref) === `ObjectPageFormActionGroup<${definitionKey}>`;
    if (hasActionGroupWrapper) {
        resolvedDefinitionKey = (0, utils_1.getDefinitionKey)(definitionOfReference.$ref);
        resolvedDefinitionOfReference = appSchema.definitions[resolvedDefinitionKey];
    }
    // create an instance of the action group in the configuration object
    importActionToConfigurationObject(importActionParams, resolvedDefinitionKey, true);
    // newly created action group instance is available in the actionsInConfig object
    const actionGroup = actionsInConfig[actionKey];
    // fill the action group with the action settings of the group action
    const { properties } = resolvedDefinitionOfReference;
    for (const propertyKey in properties) {
        const actionInGroupProperty = properties[propertyKey];
        if (actionInGroupProperty['$ref']) {
            const actionInGroupPropertyKey = (0, utils_1.getDefinitionKey)(actionInGroupProperty['$ref']);
            const actionInGroupDefinition = appSchema.definitions[actionInGroupPropertyKey];
            const key = (0, utils_1.getDefinitionKey)(actionInGroupDefinition.$ref ?? actionInGroupProperty['$ref']);
            const importGroupedActionParams = {
                ...importActionParams,
                actionsInConfig: actionGroup,
                actionKey: propertyKey
            };
            importActionToConfigurationObject(importGroupedActionParams, key);
        }
    }
    (0, common_1.removeEmptyStructure)(actionsInConfig, actionKey);
}
/**
 * Imports and updates the action menu settings in a configuration object based on a given reference definition and related action parameters.
 *
 * @param {ImportActionSettingsOfReferenceParams} importActionParams - An object containing details such as the action property definition, application schema, actions configuration, and action key. These parameters are used for resolving and importing the settings.
 * @param {Definition} definitionOfReference - The reference definition object which includes properties and other settings required to define action menus and their related configurations.
 */
function importManifestBasedActionMenuSettingsOfReference(importActionParams, definitionOfReference) {
    const { actionPropertyDefinition, appSchema, actionsInConfig, actionKey } = importActionParams;
    // create an instance of the menu action in the configuration object
    const key = (0, utils_1.getDefinitionKey)(definitionOfReference.$ref ?? actionPropertyDefinition.$ref);
    importActionToConfigurationObject(importActionParams, key, true);
    // newly created menu action instance is available in the actionsInConfig object
    const menuAction = actionsInConfig[actionKey];
    menuAction[ux_specification_types_1.PropertyName.menu] = actionPropertyDefinition[ux_specification_types_1.PropertyName.menu];
    menuAction[ux_specification_types_1.PropertyName.actions] = {};
    menuAction[ux_specification_types_1.PropertyName.text] = definitionOfReference.description;
    const actionsDefinitionKey = (0, utils_1.getDefinitionKey)(definitionOfReference?.properties?.[ux_specification_types_1.PropertyName.actions]?.$ref);
    const actionsDefinition = appSchema.definitions[actionsDefinitionKey];
    // fill the action menu with the action settings of the menu action
    const actions = actionsDefinition?.properties ?? {};
    for (const action in actions) {
        const actionInMenuActionProperty = actions[action];
        const actionInMenuActionPropertyKey = (0, utils_1.getDefinitionKey)(actionInMenuActionProperty['$ref']);
        const actionInMenuActionDefinition = appSchema.definitions[actionInMenuActionPropertyKey];
        const key = (0, utils_1.getDefinitionKey)(actionInMenuActionDefinition?.$ref ?? actionInMenuActionProperty['$ref']);
        const importMenuActionParams = {
            ...importActionParams,
            actionsInConfig: menuAction[ux_specification_types_1.PropertyName.actions],
            actionKey: action
        };
        importActionToConfigurationObject(importMenuActionParams, key);
    }
    (0, common_1.removeEmptyStructure)(actionsInConfig, actionKey);
}
/**
 * Determines the strategy type for action settings based on the provided data type and action type.
 *
 * @param {string} [dataType] - The data type associated with the action settings. Defaults to an empty string.
 * @param {string} [actionType] - The action type within the context of the action settings. Defaults to an empty string.
 * @returns {DataFieldStrategyTypes} The corresponding strategy type for the given data type and action type.
 */
function determineActionStrategyType(dataType = '', actionType = '') {
    const key = `${dataType}${actionType}`;
    return strategyTypeMap.get(key) || "com.sap.vocabularies.UI.v1.DataFieldForAction" /* UIAnnotationTypes.DataFieldForAction */;
}
/**
 * Imports the action settings of a specified reference based on the given parameters.
 *
 * @param {object} importActionParams - The parameters required for importing action settings.
 * @param {object} importActionParams.actionPropertyDefinition - The action property definition containing the $ref to the reference.
 * @param {object} importActionParams.appSchema - The application schema containing definitions and other related metadata.
 */
function importActionSettingsOfReference(importActionParams) {
    const { actionPropertyDefinition, appSchema } = importActionParams;
    const definitionKey = (0, utils_1.getDefinitionKey)(actionPropertyDefinition.$ref);
    const definitionOfReference = appSchema.definitions[definitionKey];
    const dataType = definitionOfReference?.['dataType'];
    const actionType = actionPropertyDefinition?.[ACTION_TYPE];
    const strategyType = determineActionStrategyType(dataType, actionType);
    const strategy = DataFieldStrategy_1.dataFieldStrategyContext.getStrategy(strategyType);
    strategy.importActionSettingsOfReference(importActionParams, definitionOfReference, definitionKey);
}
/**
 * Enhances the application schema with custom field definitions for Object Page sections.
 *
 * @param appSchema - The application-specific JSON schema object representing the Object Page.
 * @param v4Page - The current Object Page configuration from the manifest (V4 format).
 * @param fieldGroupInManifest - The manifest identifier for the Object Page section associated with the field definitions.
 * @param logger - Logger instance for reporting or debugging schema processing errors.
 * @param customFieldDefinitionName - The schema definition name to be used for custom fields (defaults to `CustomTableAction`).
 * @param fieldsDefinitionName - The schema definition name for the set of fields being enhanced.
 */
function addCustomFieldDefinition(appSchema, v4Page, fieldGroupInManifest, logger, customFieldDefinitionName, fieldsDefinitionName) {
    const definition = appSchema.definitions[fieldsDefinitionName];
    if (!definition) {
        return;
    }
    // Sort existing actions
    let sortedFieldsArray = sortPropertiesRecords(definition);
    const configId = fieldGroupInManifest.replace(/::/g, '#');
    const fields = v4Page.options?.settings?.controlConfiguration?.[configId]?.['fields'];
    if (fields) {
        sortedFieldsArray = addExtensionToArray(fields, sortedFieldsArray, customFieldDefinitionName, ux_specification_types_1.TemplatePropertyName.Template, logger);
    }
    // Adjust propertyIndex
    if (sortedFieldsArray) {
        for (let index = 0; index < sortedFieldsArray.length; index++) {
            sortedFieldsArray[index][1][ux_specification_types_1.SchemaTag.propertyIndex] = index;
        }
        const sortedFieldsAsObject = {};
        sortedFieldsArray.forEach(([key, value]) => (sortedFieldsAsObject[key] = value));
        appSchema.definitions[fieldsDefinitionName].properties = sortedFieldsAsObject;
    }
}
/**
 * Adds 'manifestPath' to each property in a definition based on metadata sync rules.
 *
 * @param {string} appSchema - Full schema.
 * @param {string | undefined} definitionName - The name of the definition/class whose properties are being processed.
 *   If `undefined`, the function attempts to resolve the definition name from the `$ref` property of the provided definition.
 * @param {Definition} definition - The definition object containing property schemas.
 * @param {ManifestPathParams} pathParams - Additional context required to compute manifest paths.
 * @param handleRelative - A flag indicating whether to use relative paths for definition properties.
 */
function addManifestPathsToProperties(appSchema, definitionName, definition, pathParams, handleRelative = false) {
    (0, utils_1.addManifestPathsToProperties)([v4controls, pages], appSchema, definitionName, definition, pathParams, undefined, handleRelative);
}
/**
 * Determines whether an entity set or navigation target is insertable based on its annotations.
 *
 * When navigationPath is provided:
 * 1. Checks NavigationRestrictions.RestrictedProperties on the source EntitySet
 * 2. Resolves the target EntitySet via navigationPropertyBinding
 * 3. Checks InsertRestrictions on the target EntitySet
 * 4. Both must allow insertion for the function to return true
 *
 * When navigationPath is not provided:
 * 1. Checks InsertRestrictions on the given EntitySet
 *
 * For Path-based (dynamic) Insertable values, always returns true (shows creationMode).
 *
 * @param entitySet - The source entity set (where navigation starts from, or the entity set being checked)
 * @param navigationPath - Optional navigation property path (e.g., "_Booking", "to_Items")
 * @returns A boolean indicating whether insertion is allowed. Returns true if the insertability is Path-based (dynamic) or undefined.
 */
function getInsertableForEntitySet(entitySet, navigationPath) {
    // Check NavigationRestrictions if navigation path provided
    // Default: allow if not specified (Path-based values are also treated as true)
    let navigationInsertable = true;
    if (navigationPath) {
        const navRestrictions = entitySet?.annotations?.Capabilities?.NavigationRestrictions;
        if (navRestrictions?.RestrictedProperties) {
            const restriction = navRestrictions.RestrictedProperties.find((prop) => prop.NavigationProperty?.value === navigationPath);
            if (restriction?.InsertRestrictions?.Insertable !== undefined) {
                const navInsertable = restriction.InsertRestrictions.Insertable;
                if (typeof navInsertable === 'boolean') {
                    navigationInsertable = navInsertable;
                }
            }
        }
        // Resolve target EntitySet via navigationPropertyBinding
        const targetEntitySet = entitySet?.navigationPropertyBinding?.[navigationPath];
        if (targetEntitySet && '_type' in targetEntitySet && targetEntitySet._type === 'EntitySet') {
            // Default: allow if not specified (Path-based values are also treated as true)
            let targetEntitySetInsertable = true;
            const targetRestriction = targetEntitySet.annotations?.Capabilities?.InsertRestrictions?.Insertable;
            if (typeof targetRestriction === 'boolean') {
                targetEntitySetInsertable = targetRestriction;
            }
            // Both Navigation and Target EntitySet must allow insertion
            return navigationInsertable && targetEntitySetInsertable;
        }
        // If no targetEntitySet found, only check navigation restrictions
        return navigationInsertable;
    }
    // No navigation path: check EntitySet-level InsertRestrictions
    // Default: allow if not specified (Path-based values are also treated as true)
    let entitySetInsertable = true;
    const entitySetRestriction = entitySet?.annotations?.Capabilities?.InsertRestrictions?.Insertable;
    if (typeof entitySetRestriction === 'boolean') {
        entitySetInsertable = entitySetRestriction;
    }
    return entitySetInsertable;
}
/**
 * Determines the default OData V4 table type for an entity set based on its metadata capabilities.
 *
 * The decision follows a strict priority order:
 * 1) Analytical tables take precedence when aggregate transformations are present and the
 *    analytical requirements are satisfied.
 * 2) Hierarchical tables are selected when a recursive hierarchy is defined and analytical
 *    conditions are not met.
 * 3) Responsive tables are used as a safe fallback when neither analytical nor hierarchical
 *    characteristics apply.
 *
 * The evaluation of analytical suitability differs between CAP and non-CAP services:
 * - **CAP services**: Require both service-level AND entity-level aggregation annotations.
 * - **Non-CAP services**: Service-level aggregation annotations are sufficient.
 *
 * @param entitySet The entity set whose annotations are evaluated.
 * @param isCapService Indicates whether the service is a CAP service, affecting analytical requirements.
 * @returns The default {@link TableTypeV4} to be used for the entity set.
 */
function getDefaultTableType(entitySet, isCapService) {
    if (!entitySet) {
        // Fallback when entitySet is not provided
        return FALLBACK_TABLE_TYPE;
    }
    const { hasRecursiveHierarchyForEntitySet, hasAggregateTransformations, hasAggregateTransformationsForEntitySet } = (0, project_access_1.getTableCapabilitiesByEntitySet)(entitySet);
    // Check for analytical capabilities first (highest priority)
    // Requires service-level aggregate transformations (hasAggregateTransformations) as a prerequisite
    // For CAP services: both service-level and entity-level transformations must be present
    // For non-CAP services: service-level transformations alone are sufficient
    if (hasAggregateTransformations && (hasAggregateTransformationsForEntitySet || !isCapService)) {
        return v4_1.TableTypeV4.AnalyticalTable;
    }
    // Check for hierarchical data only (no analytical data or analytical requirements not met)
    if (hasRecursiveHierarchyForEntitySet) {
        return v4_1.TableTypeV4.TreeTable;
    }
    // Default for standard entity sets without analytical or hierarchical capabilities
    return v4_1.TableTypeV4.ResponsiveTable;
}
/**
 * Determines the available table type options and recommended default type for an entity set
 * based on its metadata capabilities.
 *
 * This function analyzes the entity set's capabilities and returns both the list of suitable
 * table types and the recommended default type. The available options are restricted based
 * on the detected service characteristics:
 *
 * - **Analytical services**: AnalyticalTable, ResponsiveTable, GridTable (default: AnalyticalTable).
 * - **Hierarchical services**: TreeTable, ResponsiveTable, GridTable (default: TreeTable).
 * - **Standard services**: ResponsiveTable, GridTable (default: ResponsiveTable).
 *
 * **Version gating**: Capability-based restrictions are only applied for UI5 1.142.0+.
 * For earlier versions, all table types are available (ResponsiveTable, GridTable, AnalyticalTable, TreeTable)
 * with ResponsiveTable as the default.
 *
 * @param entitySet The entity set whose capabilities determine the available table types.
 * @param isCapService Indicates whether the service is a CAP service, affecting analytical
 *                     capability detection.
 * @param minUI5Version The minimum UI5 version from manifest. Capability-based restrictions
 *                      require UI5 1.142.0+.
 * @returns An object containing:
 *          - `options`: Array of suitable {@link TableTypeV4} values for the entity set
 *          - `defaultType`: The recommended {@link TableTypeV4} as the default choice
 * @example
 * // For an entity set with analytical capabilities (UI5 1.142+)
 * const result = getTableTypeOptions(analyticalEntitySet, true, '1.142.0');
 * // Returns: { options: [AnalyticalTable, ResponsiveTable, GridTable], defaultType: AnalyticalTable }
 *
 * // For the same entity set on UI5 1.120 (all types available, no restrictions)
 * const result = getTableTypeOptions(analyticalEntitySet, true, '1.120.0');
 * // Returns: { options: [ResponsiveTable, GridTable, AnalyticalTable, TreeTable], defaultType: ResponsiveTable }
 */
function getTableTypeOptions(entitySet, isCapService, minUI5Version) {
    // Capability-based restrictions are only available in UI5 1.142.0+
    // For earlier versions, all table types are available
    if (minUI5Version && !(0, utils_3.compareUI5Versions)(minUI5Version, exports.MIN_UI5_VERSION_FOR_TABLE_TYPE_RESTRICTIONS)) {
        return LEGACY_VERSION_TABLE_TYPE_OPTIONS;
    }
    const detectedType = getDefaultTableType(entitySet, isCapService);
    return TABLE_TYPE_OPTIONS_MAP[detectedType];
}
/**
 * Creates a virtual EntitySet for capability detection when the entitySet.entityType
 * doesn't match the actual target EntityType (e.g., for contained navigation).
 *
 * This is needed for contained navigation (ContainsTarget=true) where there is no
 * physical EntitySet in the metadata, but the target EntityType has important
 * annotations like RecursiveHierarchy.
 *
 * @param entitySet - The base entity set (may have wrong entityType for contained navigation)
 * @param targetEntityType - The actual entity type with correct annotations
 * @returns EntitySet with correct entityType for capability detection, or original if types match
 * @example
 * // For contained navigation "/ZZZC_FINANCIAL_TREE/Set"
 * const entitySetForCapabilities = ensureEntitySetWithCorrectType(
 *     entitySet,              // ZZZC_FINANCIAL_TREE (type: Parameters)
 *     lineItemFacet.entityType // ZZZC_FINANCIAL_TREEType (with RecursiveHierarchy)
 * );
 * const options = getTableTypeOptions(entitySetForCapabilities, isCapService);
 */
function ensureEntitySetWithCorrectType(entitySet, targetEntityType) {
    if (!entitySet || !targetEntityType) {
        return entitySet;
    }
    // If entityType already matches, return as-is
    if (entitySet.entityType === targetEntityType) {
        return entitySet;
    }
    // Create virtual EntitySet with correct entityType and its annotations
    return {
        ...entitySet,
        entityType: targetEntityType,
        annotations: targetEntityType.annotations || entitySet.annotations
    };
}
/**
 * Returns the appropriate fallback table type options based on UI5 version.
 * For UI5 < 1.142.0, returns all table types (legacy behavior).
 * For UI5 >= 1.142.0, returns restricted fallback options.
 *
 * @param minUI5Version - Optional minimum UI5 version
 * @returns Table type options with all types for legacy versions, or restricted options for modern versions
 */
function getFallbackTableTypeOptions(minUI5Version) {
    if (minUI5Version && !(0, utils_3.compareUI5Versions)(minUI5Version, exports.MIN_UI5_VERSION_FOR_TABLE_TYPE_RESTRICTIONS)) {
        return LEGACY_VERSION_TABLE_TYPE_OPTIONS;
    }
    return FALLBACK_TABLE_TYPE_OPTIONS;
}
/**
 * Determines table type options based on context path by resolving the entity set internally.
 *
 * This is a wrapper function that handles the complexity of entity set resolution from annotation files.
 * It parses the annotation files, resolves the entity set from the context path, and returns the
 * appropriate table type options.
 *
 * **Version gating**: Capability-based restrictions are only applied for UI5 1.142.0+.
 * For earlier versions, all table types are available.
 *
 * @param annotationFiles - Array of annotation files containing metadata
 * @param contextPath - Context path or navigation property. Can be:
 *                      - Full path starting with "/" (e.g., "/Travel", "/Travel/to_Booking")
 *                      - Navigation property without "/" (e.g., "to_Booking") - requires entitySetName
 *                      - Entity set name (e.g., "Travel")
 * @param entitySetName - Optional entity set name. Required when contextPath is a navigation property without "/"
 * @param isCapService - Indicates whether the service is a CAP service
 * @param logger - Optional logger for debugging
 * @param minUI5Version - Optional minimum UI5 version. Capability-based restrictions require UI5 1.142.0+
 * @returns An object containing:
 *          - `options`: Array of suitable {@link TableTypeV4} values for the entity set
 *          - `defaultType`: The recommended {@link TableTypeV4} as the default choice
 * @example
 * // For an analytical service with full path (UI5 1.142+)
 * const result = getTableTypeOptionsByContext(annotationFiles, "/Travel/to_Booking", undefined, true, logger, '1.142.0');
 * // Returns: { options: [AnalyticalTable, ResponsiveTable, GridTable], defaultType: AnalyticalTable }
 * @example
 * // For navigation property with entity set name (UI5 1.120)
 * const result = getTableTypeOptionsByContext(annotationFiles, "to_Booking", "Travel", true, logger, '1.120.0');
 * // Returns: { options: [ResponsiveTable, GridTable, AnalyticalTable, TreeTable], defaultType: ResponsiveTable }
 */
function getTableTypeOptionsByContext(annotationFiles, contextPath, entitySetName, isCapService, logger, minUI5Version) {
    try {
        // 1. Parse and convert annotation files to ConvertedMetadata
        const convertedMetadata = (0, utils_1.parseAndMergeAndConvert)(annotationFiles, logger);
        if (!convertedMetadata) {
            logger?.info('Could not parse annotation files, returning default table type options');
            return getFallbackTableTypeOptions(minUI5Version);
        }
        // 2. Build full context path if needed
        // If contextPath doesn't start with "/" and entitySetName is provided, combine them
        let fullContextPath = contextPath;
        if (entitySetName && contextPath && !contextPath.startsWith('/')) {
            fullContextPath = `/${entitySetName}/${contextPath}`;
        }
        // 3. Resolve entity set from context path
        const { entitySet } = (0, importProject_1.getEntityTypeEntitySetFromContextPath)(convertedMetadata, fullContextPath, entitySetName, logger);
        if (!entitySet) {
            logger?.info(`Could not resolve entity set for context path: ${fullContextPath}, returning default table type options`);
            return getFallbackTableTypeOptions(minUI5Version);
        }
        // 4. Call existing getTableTypeOptions with resolved entitySet
        return getTableTypeOptions(entitySet, isCapService, minUI5Version);
    }
    catch (error) {
        logger?.error('Error in getTableTypeOptionsByContext', error);
        // Fallback to default options on error
        return getFallbackTableTypeOptions(minUI5Version);
    }
}
/**
 * Returns the manifest path for a given page based on the provided path parameters.
 *
 * @param {ManifestPathParams} pathParams - Parameters containing page information for path resolution.
 * @returns {string | undefined} The resolved manifest path as a string, or undefined if not applicable.
 */
function getPageManifestPath(pathParams) {
    if (!Array.isArray(pathParams.pageName)) {
        const syncPath = (0, application_1.getPathForPage)([pathParams.pageName]);
        return (0, utils_1.convertSyncRulePathToJsonPath)(syncPath);
    }
}
