// @sap/ux-specification/dist/specification/src/sync/v4/export/controls/InlineEdit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isInlineEditEnabled = isInlineEditEnabled;
exports.importInlineEdit = importInlineEdit;
exports.exportInlineEdit = exportInlineEdit;
/**
 * Determines whether inline editing is enabled for a given manifest section.
 *
 * Inline editing is considered enabled if the manifest section contains an `inlineEdit`
 * object with either a `disabledFields` or `enabledFields` array.
 *
 * @param manifestSection - The manifest section object to check for inline edit configuration.
 * @returns `true` if the manifest section has inline edit enabled `false` otherwise.
 */
function isInlineEditEnabled(manifestSection) {
    const inlineEdit = typeof manifestSection === 'object' && manifestSection !== null ? manifestSection['inlineEdit'] : undefined;
    if (inlineEdit && typeof inlineEdit === 'object') {
        return Array.isArray(inlineEdit['disabledFields']) || Array.isArray(inlineEdit['enabledFields']);
    }
    return false;
}
/**
 * Imports and normalizes inline edit settings from a manifest section.
 *
 * This function extracts the inline edit configuration from a manifest section and
 * returns a normalized structure containing `enabledFields`, `disabledFields`, and
 * `connectedFields` arrays. If inline editing is not enabled, it returns `false`.
 *
 * @param manifestSection - The manifest section object containing potential inline edit settings.
 * @returns A normalized inline edit configuration object with `enabledFields`, `disabledFields`,
 *          and `connectedFields` arrays (defaulting to empty arrays if not present), or `false`
 */
function importInlineEdit(manifestSection) {
    const isEnabled = isInlineEditEnabled(manifestSection);
    if (!isEnabled) {
        return false;
    }
    const inlineEdit = typeof manifestSection === 'object' && typeof manifestSection['inlineEdit'] === 'object'
        ? manifestSection['inlineEdit']
        : {};
    return {
        enabledFields: inlineEdit['enabledFields'] ?? [],
        disabledFields: inlineEdit['disabledFields'] ?? [],
        connectedFields: inlineEdit['connectedFields'] ?? []
    };
}
/**
 * Exports inline edit configuration from a config object to a manifest section.
 *
 * This function handles three distinct scenarios for inline edit configuration:
 * 1. **Boolean `true`**: Enables inline editing with empty field arrays, initializing both `enabledFields` and `disabledFields` as empty arrays.
 * 2. **Boolean `false`**: Disables inline editing. If inline edit was previously enabled, the configuration is reset to an empty object.
 * 3. **Object configuration**: Applies the specific field configuration from the config.
 *
 * @param manifestSection - The target manifest section object where inline edit settings will be written.
 * @param configObject - The source configuration object containing the inline edit settings to be exported.
 * @param key - The property key used to access the inline edit value from `configObject`.
 * @param manifestKey - The property key under which to store the inline edit configuration in `manifestSection`.
 */
function exportInlineEdit(manifestSection, configObject, key, manifestKey) {
    const inlineEdit = configObject[key];
    if (inlineEdit === true) {
        // In this case the user has modified the inline edit settings on the way, since the import function never creates a pure boolean true value.
        // -> Set the default representation of enabled inline edit with empty arrays for enabled and disabled fields.
        manifestSection[manifestKey] = { enabledFields: [], disabledFields: [] };
    }
    else if (inlineEdit === false) {
        // Only modify the manifest when it inline edit was previously enabled, otherwise keep it as is.
        if (isInlineEditEnabled(manifestSection)) {
            delete manifestSection[manifestKey];
        }
    }
    else {
        // Ensure that properties of the current inlineEdit configuration that are not known to us stay unchanged.
        manifestSection[manifestKey] = structuredClone(manifestSection[manifestKey] || {});
        manifestSection[manifestKey]['enabledFields'] = inlineEdit.enabledFields;
        manifestSection[manifestKey]['disabledFields'] = inlineEdit.disabledFields;
        if (inlineEdit.connectedFields.length > 0) {
            manifestSection[manifestKey]['connectedFields'] = inlineEdit.connectedFields;
        }
        else if (Array.isArray(manifestSection[manifestKey]['connectedFields'])) {
            delete manifestSection[manifestKey]['connectedFields'];
        }
    }
}
