// @sap/ux-specification/dist/specification/src/sync/v4/export/controls/ObjectPageCustomSection.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectPageCustomSubSectionFragment = exports.ObjectPageEmbeddedComponentSection = exports.ObjectPageCustomSectionFragment = exports.ObjectPageCustomSectionActions = void 0;
const ux_specification_types_1 = require("@sap/ux-specification-types");
const decorators_1 = require("../../../common/decoration/decorators");
const application_1 = require("../../application");
class ObjectPageCustomSectionActions {
}
exports.ObjectPageCustomSectionActions = ObjectPageCustomSectionActions;
class ObjectPageCustomSectionFragment {
}
exports.ObjectPageCustomSectionFragment = ObjectPageCustomSectionFragment;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId) => {
                return (0, application_1.getCustomSectionPath)(pageName, manifest, sectionId, 'position');
            },
            key: 'anchor'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSectionFragment.prototype, "relatedFacet", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId) => (0, application_1.getCustomSectionPath)(pageName, manifest, sectionId, 'position'),
            key: 'placement'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSectionFragment.prototype, "relativePosition", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSectionPath
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSectionFragment.prototype, "title", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSectionPath,
            import: (manifestSection) => {
                // During import, we need to consider both `name` and `template` to correctly fill our config and to be upgrade safe
                return manifestSection.name || manifestSection.template;
            },
            key: 'template' // key: 'name' deprecated since '1.86'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSectionFragment.prototype, "fragmentName", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSectionPath,
            import: false,
            export: false
        },
        xml: (config) => {
            return config['fragmentName'];
        }
    }),
    (0, decorators_1.validity)({
        since: '1.96.0'
    })
], ObjectPageCustomSectionFragment.prototype, "controls", void 0);
/**
 * Builds a syncRule for a property nested under `embeddedComponent/settings` in the manifest.
 *
 * @param key - The manifest key within embeddedComponent/settings
 * @returns SyncRule descriptor with the embeddedComponent/settings path and the given key
 */
function embeddedSettingsSyncRule(key) {
    return {
        manifest: {
            path: (pageName, manifest, sectionId) => (0, application_1.getCustomSectionPath)(pageName, manifest, sectionId, `${ux_specification_types_1.TemplatePropertyName.EmbeddedComponent}/settings`),
            key
        }
    };
}
class ObjectPageEmbeddedComponentSection extends ObjectPageCustomSectionFragment {
}
exports.ObjectPageEmbeddedComponentSection = ObjectPageEmbeddedComponentSection;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId) => (0, application_1.getCustomSectionPath)(pageName, manifest, sectionId, ux_specification_types_1.TemplatePropertyName.EmbeddedComponent),
            key: 'name',
            export: (manifestSection, configObject) => {
                const embeddedComponent = configObject?.embeddedComponent;
                if (embeddedComponent?.name) {
                    manifestSection['name'] = embeddedComponent.name;
                }
            },
            import: (manifestSection) => {
                const name = manifestSection?.['name'];
                return name ? { name } : undefined;
            }
        }
    }),
    (0, decorators_1.validity)({
        since: '1.136.0'
    }),
    (0, decorators_1.hide)(true)
], ObjectPageEmbeddedComponentSection.prototype, "embeddedComponent", void 0);
__decorate([
    (0, decorators_1.syncRule)(embeddedSettingsSyncRule('sapObjectNodeType')),
    (0, decorators_1.validity)({
        since: '1.136.0'
    })
], ObjectPageEmbeddedComponentSection.prototype, "sapObjectNodeType", void 0);
__decorate([
    (0, decorators_1.syncRule)(embeddedSettingsSyncRule('linkedObjectKey')),
    (0, decorators_1.validity)({
        since: '1.136.0'
    })
], ObjectPageEmbeddedComponentSection.prototype, "linkedObjectKey", void 0);
__decorate([
    (0, decorators_1.syncRule)(embeddedSettingsSyncRule('mode')),
    (0, decorators_1.validity)({ since: '1.136.0' }),
    (0, decorators_1.hide)(true)
], ObjectPageEmbeddedComponentSection.prototype, "mode", void 0);
__decorate([
    (0, decorators_1.syncRule)(embeddedSettingsSyncRule('fioriId')),
    (0, decorators_1.validity)({ since: '1.136.0' }),
    (0, decorators_1.hide)(true)
], ObjectPageEmbeddedComponentSection.prototype, "fioriId", void 0);
__decorate([
    (0, decorators_1.syncRule)(embeddedSettingsSyncRule('appId')),
    (0, decorators_1.validity)({ since: '1.136.0' }),
    (0, decorators_1.hide)(true)
], ObjectPageEmbeddedComponentSection.prototype, "appId", void 0);
__decorate([
    (0, decorators_1.syncRule)(embeddedSettingsSyncRule('objectPageId')),
    (0, decorators_1.validity)({ since: '1.136.0' }),
    (0, decorators_1.hide)(true)
], ObjectPageEmbeddedComponentSection.prototype, "objectPageId", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSectionPath,
            import: false,
            export: false
        }
    })
], ObjectPageEmbeddedComponentSection.prototype, "fragmentName", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSectionPath,
            import: false,
            export: false
        },
        xml: undefined
    })
], ObjectPageEmbeddedComponentSection.prototype, "controls", void 0);
class ObjectPageCustomSubSectionFragment extends ObjectPageCustomSectionFragment {
}
exports.ObjectPageCustomSubSectionFragment = ObjectPageCustomSubSectionFragment;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (...args) => {
                const resolvedPath = application_1.getCustomSubSectionPath.apply({}, args);
                return `${resolvedPath}/position`;
            },
            key: 'anchor'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSubSectionFragment.prototype, "relatedFacet", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (...args) => {
                const resolvedPath = application_1.getCustomSubSectionPath.apply({}, args);
                return `${resolvedPath}/position`;
            },
            key: 'placement'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSubSectionFragment.prototype, "relativePosition", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSubSectionPath
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSubSectionFragment.prototype, "title", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSubSectionPath,
            import: (manifestSection) => {
                // During import, we need to consider both `name` and `template` to correctly fill our config and to be upgrade safe
                return manifestSection.name || manifestSection.template;
            },
            key: 'template' // key: 'name' deprecated since '1.86'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSubSectionFragment.prototype, "fragmentName", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSubSectionPath,
            import: false,
            export: false
        },
        xml: (config) => {
            return config['fragmentName'];
        }
    }),
    (0, decorators_1.validity)({
        since: '1.96.0'
    })
], ObjectPageCustomSubSectionFragment.prototype, "controls", void 0);
